// @openai/agents-core/dist/sandbox/entries/factories.mjs
export function dir(args = {}) {
    return {
        type: 'dir',
        ...args,
    };
}
export function file(args) {
    return {
        type: 'file',
        ...args,
    };
}
export function localFile(args) {
    return {
        type: 'local_file',
        ...args,
    };
}
export function localDir(args) {
    return {
        type: 'local_dir',
        ...args,
    };
}
export function gitRepo(args) {
    return {
        type: 'git_repo',
        ...args,
    };
}
export function mount(args = {}) {
    return {
        type: 'mount',
        ...args,
    };
}
export function s3Mount(args) {
    return {
        type: 's3_mount',
        ...args,
    };
}
export function gcsMount(args) {
    return {
        type: 'gcs_mount',
        ...args,
    };
}
export function r2Mount(args) {
    return {
        type: 'r2_mount',
        ...args,
    };
}
export function azureBlobMount(args) {
    return {
        type: 'azure_blob_mount',
        ...args,
    };
}
export function boxMount(args = {}) {
    return {
        type: 'box_mount',
        ...args,
    };
}
export function s3FilesMount(args) {
    return {
        type: 's3_files_mount',
        ...args,
    };
}
export function inContainerMountStrategy(args = {}) {
    return {
        type: 'in_container',
        ...args,
    };
}
export function dockerVolumeMountStrategy(args = {}) {
    return {
        type: 'docker_volume',
        ...args,
    };
}
export function localBindMountStrategy(args = {}) {
    return {
        type: 'local_bind',
        ...args,
    };
}
export function mountPattern(args) {
    return args;
}
