// @openai/agents-core/dist/sandbox/entries/factories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dir = dir;
exports.file = file;
exports.localFile = localFile;
exports.localDir = localDir;
exports.gitRepo = gitRepo;
exports.mount = mount;
exports.s3Mount = s3Mount;
exports.gcsMount = gcsMount;
exports.r2Mount = r2Mount;
exports.azureBlobMount = azureBlobMount;
exports.boxMount = boxMount;
exports.s3FilesMount = s3FilesMount;
exports.inContainerMountStrategy = inContainerMountStrategy;
exports.dockerVolumeMountStrategy = dockerVolumeMountStrategy;
exports.localBindMountStrategy = localBindMountStrategy;
exports.mountPattern = mountPattern;
function dir(args = {}) {
    return {
        type: 'dir',
        ...args,
    };
}
function file(args) {
    return {
        type: 'file',
        ...args,
    };
}
function localFile(args) {
    return {
        type: 'local_file',
        ...args,
    };
}
function localDir(args) {
    return {
        type: 'local_dir',
        ...args,
    };
}
function gitRepo(args) {
    return {
        type: 'git_repo',
        ...args,
    };
}
function mount(args = {}) {
    return {
        type: 'mount',
        ...args,
    };
}
function s3Mount(args) {
    return {
        type: 's3_mount',
        ...args,
    };
}
function gcsMount(args) {
    return {
        type: 'gcs_mount',
        ...args,
    };
}
function r2Mount(args) {
    return {
        type: 'r2_mount',
        ...args,
    };
}
function azureBlobMount(args) {
    return {
        type: 'azure_blob_mount',
        ...args,
    };
}
function boxMount(args = {}) {
    return {
        type: 'box_mount',
        ...args,
    };
}
function s3FilesMount(args) {
    return {
        type: 's3_files_mount',
        ...args,
    };
}
function inContainerMountStrategy(args = {}) {
    return {
        type: 'in_container',
        ...args,
    };
}
function dockerVolumeMountStrategy(args = {}) {
    return {
        type: 'docker_volume',
        ...args,
    };
}
function localBindMountStrategy(args = {}) {
    return {
        type: 'local_bind',
        ...args,
    };
}
function mountPattern(args) {
    return args;
}
