// @openai/agents-core/dist/runner/sandbox.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareSandboxInterruptedTurnResume = prepareSandboxInterruptedTurnResume;
exports.finalizeSandboxRuntime = finalizeSandboxRuntime;
exports.rehydrateInterruptedTurnExecutionTools = rehydrateInterruptedTurnExecutionTools;
exports.isSandboxRuntimeAgent = isSandboxRuntimeAgent;
const errors_1 = require("../errors.js");
const logger_1 = __importDefault(require("../logger.js"));
const runState_1 = require("../runState.js");
const agentKeys_1 = require("../sandbox/runtime/agentKeys.js");
const toolRehydration_1 = require("../sandbox/runtime/toolRehydration.js");
const tool_1 = require("../tool.js");
const context_1 = require("../tracing/context.js");
const modelPreparation_1 = require("./modelPreparation.js");
async function prepareSandboxInterruptedTurnResume(args) {
    const { startingAgent, state, sandboxRuntime, runConfigModel } = args;
    logger_1.default.debug('Continuing from interruption');
    if (!state._lastTurnResponse || !state._lastProcessedResponse) {
        throw new errors_1.UserError('No model response found in previous state', state);
    }
    const requiresExecutionToolRehydration = (0, toolRehydration_1.processedResponseRequiresExecutionToolRehydration)(state._lastProcessedResponse);
    if (!(0, agentKeys_1.isSandboxAgent)(state._currentAgent) &&
        !requiresExecutionToolRehydration) {
        return;
    }
    const resumedPreservedSessions = await sandboxRuntime.adoptPreservedOwnedSessions();
    if (resumedPreservedSessions || requiresExecutionToolRehydration) {
        await rehydrateInterruptedTurnExecutionTools({
            startingAgent,
            state,
            sandboxRuntime,
            runConfigModel,
            force: resumedPreservedSessions,
        });
    }
}
async function finalizeSandboxRuntime(args) {
    const { state, sandboxRuntime, preserveSessionsForInterruption, runError, groupId, memoryContext, runAgent, } = args;
    if (!preserveSessionsForInterruption) {
        try {
            await (0, tool_1.disposeResolvedComputers)({ runContext: state._context });
        }
        catch (error) {
            logger_1.default.warn(`Failed to dispose computers after run: ${error}`);
        }
    }
    await sandboxRuntime.enqueueMemoryGeneration(state, {
        exception: runError,
        groupId,
        inputOverride: memoryContext?.inputOverride?.(),
        sdkSessionId: memoryContext?.sdkSessionId,
        runAgent: async (agent, input, runOptions) => await runAgent(agent, input, runOptions),
    });
    try {
        await sandboxRuntime.cleanup(state, {
            preserveOwnedSessions: preserveSessionsForInterruption,
        });
    }
    finally {
        if (state._currentAgentSpan) {
            try {
                if (!preserveSessionsForInterruption) {
                    state._currentAgentSpan.end();
                }
            }
            finally {
                (0, context_1.resetCurrentSpan)();
            }
        }
    }
}
async function rehydrateInterruptedTurnExecutionTools(args) {
    const { startingAgent, state, sandboxRuntime, runConfigModel, force } = args;
    if (!force &&
        !(0, toolRehydration_1.processedResponseRequiresExecutionToolRehydration)(state._lastProcessedResponse)) {
        return;
    }
    const preparedSandboxAgent = await sandboxRuntime.prepareAgent({
        currentAgent: state._currentAgent,
        turnInput: [],
        runConfigModel,
    });
    const artifacts = await (0, modelPreparation_1.prepareAgentArtifacts)(state, preparedSandboxAgent.executionAgent);
    await (0, runState_1.rehydrateProcessedResponseTools)(startingAgent, state, artifacts.tools);
}
function isSandboxRuntimeAgent(agent) {
    return (0, agentKeys_1.isSandboxAgent)(agent);
}
