// @openai/agents-core/dist/runner/streamReconciliation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStreamAbortReconciliationState = createStreamAbortReconciliationState;
exports.recordStreamEventForAbortReconciliation = recordStreamEventForAbortReconciliation;
exports.buildAbortReconciliationInput = buildAbortReconciliationInput;
exports.getAbortReconciliationPreviousResponseId = getAbortReconciliationPreviousResponseId;
exports.shouldReconcileStreamAbort = shouldReconcileStreamAbort;
exports.markAbortReconciliationComplete = markAbortReconciliationComplete;
function createStreamAbortReconciliationState() {
    return {
        pendingFunctionCalls: new Map(),
    };
}
function recordStreamEventForAbortReconciliation(state, event) {
    if (event.type === 'response_done') {
        state.pendingFunctionCalls.clear();
        state.responseId = event.response.id;
        return;
    }
    if (event.type !== 'model' || !isRecord(event.event)) {
        return;
    }
    const rawEvent = event.event;
    if (rawEvent.type === 'response.created' &&
        isRecord(rawEvent.response) &&
        typeof rawEvent.response.id === 'string') {
        state.responseId = rawEvent.response.id;
        return;
    }
    if (rawEvent.type !== 'response.output_item.done' ||
        !isRecord(rawEvent.item)) {
        return;
    }
    const item = rawEvent.item;
    if (item.type === 'function_call' && typeof item.call_id === 'string') {
        state.pendingFunctionCalls.set(item.call_id, {
            callId: item.call_id,
            name: typeof item.name === 'string' ? item.name : item.call_id,
            ...(typeof item.namespace === 'string'
                ? { namespace: item.namespace }
                : {}),
        });
        return;
    }
    if (item.type === 'function_call_output' &&
        typeof item.call_id === 'string') {
        state.pendingFunctionCalls.delete(item.call_id);
    }
}
function buildAbortReconciliationInput(state) {
    return Array.from(state.pendingFunctionCalls.values(), (toolCall) => ({
        type: 'function_call_result',
        name: toolCall.name,
        ...(typeof toolCall.namespace === 'string'
            ? { namespace: toolCall.namespace }
            : {}),
        callId: toolCall.callId,
        status: 'incomplete',
        output: { type: 'text', text: 'aborted' },
    }));
}
function getAbortReconciliationPreviousResponseId(state, request) {
    if (request.conversationId) {
        return request.previousResponseId;
    }
    return state.responseId ?? request.previousResponseId;
}
function shouldReconcileStreamAbort(state) {
    return state.pendingFunctionCalls.size > 0;
}
function markAbortReconciliationComplete(state, response) {
    state.pendingFunctionCalls.clear();
    if (response?.responseId) {
        state.responseId = response.responseId;
    }
}
function isRecord(value) {
    return typeof value === 'object' && value !== null;
}
