// @openai/agents-core/dist/mcp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultMCPToolCacheKey = exports.MCPServerSSE = exports.MCPServerStreamableHttp = exports.MCPServerStdio = exports.MCPTool = exports.BaseMCPServerSSE = exports.BaseMCPServerStreamableHttp = exports.BaseMCPServerStdio = exports.DEFAULT_SSE_MCP_CLIENT_LOGGER_NAME = exports.DEFAULT_STREAMABLE_HTTP_MCP_CLIENT_LOGGER_NAME = exports.DEFAULT_STDIO_MCP_CLIENT_LOGGER_NAME = void 0;
exports.invalidateServerToolsCache = invalidateServerToolsCache;
exports.getAllMcpTools = getAllMcpTools;
exports.mcpToFunctionTool = mcpToFunctionTool;
exports.attachCallToolResultMetadata = attachCallToolResultMetadata;
const tool_1 = require("./tool.js");
const errors_1 = require("./errors.js");
const _shims_1 = require("@openai/agents-core/_shims");
const tracing_1 = require("./tracing/index.js");
const logger_1 = require("./logger.js");
const debug_1 = __importDefault(require("debug"));
const zod_1 = require("zod");
const customData_1 = require("./utils/customData.js");
exports.DEFAULT_STDIO_MCP_CLIENT_LOGGER_NAME = 'openai-agents:stdio-mcp-client';
exports.DEFAULT_STREAMABLE_HTTP_MCP_CLIENT_LOGGER_NAME = 'openai-agents:streamable-http-mcp-client';
exports.DEFAULT_SSE_MCP_CLIENT_LOGGER_NAME = 'openai-agents:sse-mcp-client';
const MCP_FUNCTION_TOOL_NAME_MAX_LENGTH = 64;
const MCP_FUNCTION_TOOL_HASH_LENGTH = 8;
class BaseMCPServerStdio {
    cacheToolsList;
    _cachedTools = undefined;
    toolFilter;
    toolMetaResolver;
    customDataExtractor;
    useStructuredContent;
    errorFunction;
    logger;
    constructor(options) {
        this.logger =
            options.logger ?? (0, logger_1.getLogger)(exports.DEFAULT_STDIO_MCP_CLIENT_LOGGER_NAME);
        this.cacheToolsList = options.cacheToolsList ?? false;
        this.toolFilter = options.toolFilter;
        this.toolMetaResolver = options.toolMetaResolver;
        this.customDataExtractor = options.customDataExtractor;
        this.useStructuredContent = options.useStructuredContent;
        this.errorFunction = options.errorFunction;
    }
    /**
     * Logs a debug message when debug logging is enabled.
     * @param buildMessage A function that returns the message to log.
     */
    debugLog(buildMessage) {
        if (debug_1.default.enabled(this.logger.namespace)) {
            // only when this is true, the function to build the string is called
            this.logger.debug(buildMessage());
        }
    }
}
exports.BaseMCPServerStdio = BaseMCPServerStdio;
class BaseMCPServerStreamableHttp {
    cacheToolsList;
    _cachedTools = undefined;
    toolFilter;
    toolMetaResolver;
    customDataExtractor;
    useStructuredContent;
    errorFunction;
    logger;
    constructor(options) {
        this.logger =
            options.logger ??
                (0, logger_1.getLogger)(exports.DEFAULT_STREAMABLE_HTTP_MCP_CLIENT_LOGGER_NAME);
        this.cacheToolsList = options.cacheToolsList ?? false;
        this.toolFilter = options.toolFilter;
        this.toolMetaResolver = options.toolMetaResolver;
        this.customDataExtractor = options.customDataExtractor;
        this.useStructuredContent = options.useStructuredContent;
        this.errorFunction = options.errorFunction;
    }
    /**
     * Logs a debug message when debug logging is enabled.
     * @param buildMessage A function that returns the message to log.
     */
    debugLog(buildMessage) {
        if (debug_1.default.enabled(this.logger.namespace)) {
            // only when this is true, the function to build the string is called
            this.logger.debug(buildMessage());
        }
    }
}
exports.BaseMCPServerStreamableHttp = BaseMCPServerStreamableHttp;
class BaseMCPServerSSE {
    cacheToolsList;
    _cachedTools = undefined;
    toolFilter;
    toolMetaResolver;
    customDataExtractor;
    useStructuredContent;
    errorFunction;
    logger;
    constructor(options) {
        this.logger =
            options.logger ?? (0, logger_1.getLogger)(exports.DEFAULT_SSE_MCP_CLIENT_LOGGER_NAME);
        this.cacheToolsList = options.cacheToolsList ?? false;
        this.toolFilter = options.toolFilter;
        this.toolMetaResolver = options.toolMetaResolver;
        this.customDataExtractor = options.customDataExtractor;
        this.useStructuredContent = options.useStructuredContent;
        this.errorFunction = options.errorFunction;
    }
    /**
     * Logs a debug message when debug logging is enabled.
     * @param buildMessage A function that returns the message to log.
     */
    debugLog(buildMessage) {
        if (debug_1.default.enabled(this.logger.namespace)) {
            // only when this is true, the function to build the string is called
            this.logger.debug(buildMessage());
        }
    }
}
exports.BaseMCPServerSSE = BaseMCPServerSSE;
/**
 * Minimum MCP tool data definition.
 * This type definition does not intend to cover all possible properties.
 * It supports the properties that are used in this SDK.
 */
exports.MCPTool = zod_1.z.object({
    name: zod_1.z.string(),
    description: zod_1.z.string().optional(),
    inputSchema: zod_1.z.object({
        type: zod_1.z.literal('object'),
        properties: zod_1.z.record(zod_1.z.string(), zod_1.z.any()),
        required: zod_1.z.array(zod_1.z.string()),
        additionalProperties: zod_1.z.boolean(),
    }),
});
/**
 * Public interface of an MCP server that provides tools.
 * You can use this class to pass MCP server settings to your agent.
 */
class MCPServerStdio extends BaseMCPServerStdio {
    underlying;
    constructor(options) {
        super(options);
        this.underlying = new _shims_1.MCPServerStdio(options);
    }
    get name() {
        return this.underlying.name;
    }
    connect() {
        return this.underlying.connect();
    }
    close() {
        return this.underlying.close();
    }
    async listTools() {
        if (this.cacheToolsList && this._cachedTools) {
            return this._cachedTools;
        }
        const tools = await this.underlying.listTools();
        if (this.cacheToolsList) {
            this._cachedTools = tools;
        }
        return tools;
    }
    async callTool(toolName, args, meta) {
        return (await this.callToolResult(toolName, args, meta)).content;
    }
    callToolResult(toolName, args, meta) {
        return this.underlying.callToolResult(toolName, args, meta);
    }
    listResources(params) {
        return this.underlying.listResources(params);
    }
    listResourceTemplates(params) {
        return this.underlying.listResourceTemplates(params);
    }
    readResource(uri) {
        return this.underlying.readResource(uri);
    }
    invalidateToolsCache() {
        return this.underlying.invalidateToolsCache();
    }
}
exports.MCPServerStdio = MCPServerStdio;
class MCPServerStreamableHttp extends BaseMCPServerStreamableHttp {
    underlying;
    _cachedToolsSessionId = undefined;
    constructor(options) {
        super(options);
        this.underlying = new _shims_1.MCPServerStreamableHttp(options);
    }
    clearLocalToolsCache() {
        this._cachedTools = undefined;
        this._cachedToolsSessionId = undefined;
    }
    get name() {
        return this.underlying.name;
    }
    get sessionId() {
        return this.underlying.sessionId;
    }
    async connect() {
        this.clearLocalToolsCache();
        await this.underlying.connect();
    }
    async close() {
        this.clearLocalToolsCache();
        await this.underlying.close();
    }
    async listTools() {
        const sessionId = this.sessionId;
        if (sessionId === undefined) {
            this.clearLocalToolsCache();
            await this.underlying.invalidateToolsCache();
            return this.underlying.listTools();
        }
        if (this.cacheToolsList &&
            this._cachedTools &&
            this._cachedToolsSessionId === sessionId) {
            return this._cachedTools;
        }
        const tools = await this.underlying.listTools();
        if (this.cacheToolsList) {
            this._cachedTools = tools;
            this._cachedToolsSessionId = sessionId;
        }
        return tools;
    }
    async callTool(toolName, args, meta) {
        return (await this.callToolResult(toolName, args, meta)).content;
    }
    async callToolResult(toolName, args, meta) {
        const previousSessionId = this.sessionId;
        try {
            return await this.underlying.callToolResult(toolName, args, meta);
        }
        finally {
            if (previousSessionId !== this.sessionId) {
                this.clearLocalToolsCache();
            }
        }
    }
    listResources(params) {
        return this.underlying.listResources(params);
    }
    listResourceTemplates(params) {
        return this.underlying.listResourceTemplates(params);
    }
    readResource(uri) {
        return this.underlying.readResource(uri);
    }
    async invalidateToolsCache() {
        this.clearLocalToolsCache();
        await this.underlying.invalidateToolsCache();
    }
}
exports.MCPServerStreamableHttp = MCPServerStreamableHttp;
class MCPServerSSE extends BaseMCPServerSSE {
    underlying;
    constructor(options) {
        super(options);
        this.underlying = new _shims_1.MCPServerSSE(options);
    }
    get name() {
        return this.underlying.name;
    }
    connect() {
        return this.underlying.connect();
    }
    close() {
        return this.underlying.close();
    }
    async listTools() {
        if (this.cacheToolsList && this._cachedTools) {
            return this._cachedTools;
        }
        const tools = await this.underlying.listTools();
        if (this.cacheToolsList) {
            this._cachedTools = tools;
        }
        return tools;
    }
    async callTool(toolName, args, meta) {
        return (await this.callToolResult(toolName, args, meta)).content;
    }
    callToolResult(toolName, args, meta) {
        return this.underlying.callToolResult(toolName, args, meta);
    }
    listResources(params) {
        return this.underlying.listResources(params);
    }
    listResourceTemplates(params) {
        return this.underlying.listResourceTemplates(params);
    }
    readResource(uri) {
        return this.underlying.readResource(uri);
    }
    invalidateToolsCache() {
        return this.underlying.invalidateToolsCache();
    }
}
exports.MCPServerSSE = MCPServerSSE;
/**
 * Fetches and flattens all tools from multiple MCP servers.
 * Logs and skips any servers that fail to respond.
 */
const _cachedTools = {};
const _cachedToolKeysByServer = {};
/**
 * Remove cached tools for the given server so the next lookup fetches fresh data.
 *
 * @param serverName - Name of the MCP server whose cache should be cleared.
 */
async function invalidateServerToolsCache(serverName) {
    const cachedKeys = _cachedToolKeysByServer[serverName];
    if (cachedKeys) {
        for (const cacheKey of cachedKeys) {
            delete _cachedTools[cacheKey];
        }
        delete _cachedToolKeysByServer[serverName];
        return;
    }
    delete _cachedTools[serverName];
    for (const cacheKey of Object.keys(_cachedTools)) {
        if (cacheKey.startsWith(`${serverName}:`)) {
            delete _cachedTools[cacheKey];
        }
    }
}
/**
 * Default cache key generator for MCP tools.
 * Uses server name, or server+agent if using callable filter.
 */
const defaultMCPToolCacheKey = ({ server, agent, }) => {
    if (server.toolFilter && typeof server.toolFilter === 'function' && agent) {
        return `${server.name}:${agent.name}`;
    }
    return server.name;
};
exports.defaultMCPToolCacheKey = defaultMCPToolCacheKey;
/**
 * Fetches and filters raw MCP tools from a single MCP server.
 */
async function getMcpToolsFromServer({ server, runContext, agent, generateMCPToolCacheKey, }) {
    const cacheKey = (generateMCPToolCacheKey || exports.defaultMCPToolCacheKey)({
        server,
        agent,
        runContext,
    });
    // Use cache key generator injected from the outside, or the default if absent.
    if (server.cacheToolsList && _cachedTools[cacheKey]) {
        return _cachedTools[cacheKey];
    }
    const listToolsForServer = async (span) => {
        const fetchedMcpTools = await server.listTools();
        let mcpTools = fetchedMcpTools;
        if (runContext && agent) {
            const context = { runContext, agent, serverName: server.name };
            const filteredTools = [];
            for (const tool of fetchedMcpTools) {
                const filter = server.toolFilter;
                if (filter) {
                    if (typeof filter === 'function') {
                        const filtered = await filter(context, tool);
                        if (!filtered) {
                            logger_1.logger.debug(`MCP Tool (server: ${server.name}, tool: ${tool.name}) is blocked by the callable filter.`);
                            continue;
                        }
                    }
                    else {
                        const allowedToolNames = filter.allowedToolNames ?? [];
                        const blockedToolNames = filter.blockedToolNames ?? [];
                        if (allowedToolNames.length > 0 || blockedToolNames.length > 0) {
                            const allowed = allowedToolNames.length > 0
                                ? allowedToolNames.includes(tool.name)
                                : true;
                            const blocked = blockedToolNames.length > 0
                                ? blockedToolNames.includes(tool.name)
                                : false;
                            if (!allowed || blocked) {
                                if (blocked) {
                                    logger_1.logger.debug(`MCP Tool (server: ${server.name}, tool: ${tool.name}) is blocked by the static filter.`);
                                }
                                else if (!allowed) {
                                    logger_1.logger.debug(`MCP Tool (server: ${server.name}, tool: ${tool.name}) is not allowed by the static filter.`);
                                }
                                continue;
                            }
                        }
                    }
                }
                filteredTools.push(tool);
            }
            mcpTools = filteredTools;
        }
        if (span) {
            span.spanData.result = mcpTools.map((t) => t.name);
        }
        // Cache store
        if (server.cacheToolsList) {
            _cachedTools[cacheKey] = mcpTools;
            if (!_cachedToolKeysByServer[server.name]) {
                _cachedToolKeysByServer[server.name] = new Set();
            }
            _cachedToolKeysByServer[server.name].add(cacheKey);
        }
        return mcpTools;
    };
    if (!(0, tracing_1.getCurrentTrace)()) {
        return listToolsForServer();
    }
    return (0, tracing_1.withMCPListToolsSpan)(listToolsForServer, {
        data: { server: server.name },
    });
}
function convertMcpToolsToFunctionTools({ mcpTools, server, convertSchemasToStrict, toolNameOverrides, errorFunction, }) {
    return mcpTools.map((mcpTool, index) => mcpToFunctionTool(mcpTool, server, convertSchemasToStrict, {
        toolNameOverride: toolNameOverrides?.[index],
        errorFunction,
    }));
}
/**
 * Fetches all function tools from a single MCP server.
 */
async function getFunctionToolsFromServer({ server, convertSchemasToStrict, runContext, agent, generateMCPToolCacheKey, errorFunction, }) {
    const mcpTools = await getMcpToolsFromServer({
        server,
        runContext,
        agent,
        generateMCPToolCacheKey,
    });
    return convertMcpToolsToFunctionTools({
        mcpTools,
        server,
        convertSchemasToStrict,
        errorFunction,
    });
}
/**
 * Returns all MCP tools from the provided servers, using the function tool conversion.
 * If runContext and agent are provided, callable tool filters will be applied.
 */
async function getAllMcpTools(mcpServersOrOpts, runContext, agent, convertSchemasToStrict = false) {
    const opts = Array.isArray(mcpServersOrOpts)
        ? {
            mcpServers: mcpServersOrOpts,
            runContext,
            agent,
            convertSchemasToStrict,
        }
        : mcpServersOrOpts;
    const { mcpServers, convertSchemasToStrict: convertSchemasToStrictFromOpts = false, runContext: runContextFromOpts, agent: agentFromOpts, generateMCPToolCacheKey, errorFunction, includeServerInToolNames = false, reservedToolNames, } = opts;
    const allTools = [];
    const toolNames = new Set();
    if (includeServerInToolNames) {
        const serverToolBatches = await Promise.all(mcpServers.map(async (server, serverIndex) => ({
            server,
            serverIndex,
            mcpTools: await getMcpToolsFromServer({
                server,
                runContext: runContextFromOpts,
                agent: agentFromOpts,
                generateMCPToolCacheKey,
            }),
        })));
        const toolNameOverrides = buildPrefixedToolNameOverrides(serverToolBatches, new Set(reservedToolNames ?? []));
        for (const { server, serverIndex, mcpTools } of serverToolBatches) {
            const serverTools = convertMcpToolsToFunctionTools({
                mcpTools,
                server,
                convertSchemasToStrict: convertSchemasToStrictFromOpts,
                errorFunction,
                toolNameOverrides: mcpTools.map((_, toolIndex) => toolNameOverrides.get(getToolNameOverrideKey(serverIndex, toolIndex))),
            });
            const serverToolNames = new Set(serverTools.map((t) => t.name));
            const intersection = [...serverToolNames]
                .filter((n) => toolNames.has(n))
                .sort();
            if (intersection.length > 0) {
                throw new errors_1.UserError(`Duplicate tool names found across MCP servers: ${intersection.join(', ')}`);
            }
            for (const t of serverTools) {
                toolNames.add(t.name);
                allTools.push(t);
            }
        }
        return allTools;
    }
    for (const server of mcpServers) {
        const serverTools = await getFunctionToolsFromServer({
            server,
            convertSchemasToStrict: convertSchemasToStrictFromOpts,
            runContext: runContextFromOpts,
            agent: agentFromOpts,
            generateMCPToolCacheKey,
            errorFunction,
        });
        const serverToolNames = new Set(serverTools.map((t) => t.name));
        const intersection = [...serverToolNames]
            .filter((n) => toolNames.has(n))
            .sort();
        if (intersection.length > 0) {
            throw new errors_1.UserError(`Duplicate tool names found across MCP servers: ${intersection.join(', ')}`);
        }
        for (const t of serverTools) {
            toolNames.add(t.name);
            allTools.push(t);
        }
    }
    return allTools;
}
function getToolNameOverrideKey(serverIndex, toolIndex) {
    return `${serverIndex}:${toolIndex}`;
}
function getSafeToolNamePart(value, fallback) {
    const safe = Array.from(value)
        .map((char) => {
        const code = char.charCodeAt(0);
        return code <= 127 && /[A-Za-z0-9_-]/.test(char) ? char : '_';
    })
        .join('')
        .replace(/^[_-]+|[_-]+$/g, '');
    return safe || fallback;
}
function getUtf8Bytes(value) {
    const encoded = encodeURIComponent(value);
    const bytes = [];
    for (let index = 0; index < encoded.length; index += 1) {
        if (encoded[index] === '%') {
            bytes.push(Number.parseInt(encoded.slice(index + 1, index + 3), 16));
            index += 2;
        }
        else {
            bytes.push(encoded.charCodeAt(index));
        }
    }
    return bytes;
}
function rotateLeft(value, bits) {
    return ((value << bits) | (value >>> (32 - bits))) >>> 0;
}
function getSha1Hex(value) {
    const bytes = getUtf8Bytes(value);
    const bitLength = bytes.length * 8;
    bytes.push(0x80);
    while (bytes.length % 64 !== 56) {
        bytes.push(0);
    }
    for (let shift = 56; shift >= 0; shift -= 8) {
        bytes.push(Math.floor(bitLength / 2 ** shift) & 0xff);
    }
    let h0 = 0x67452301;
    let h1 = 0xefcdab89;
    let h2 = 0x98badcfe;
    let h3 = 0x10325476;
    let h4 = 0xc3d2e1f0;
    for (let offset = 0; offset < bytes.length; offset += 64) {
        const words = new Array(80).fill(0);
        for (let index = 0; index < 16; index += 1) {
            const byteOffset = offset + index * 4;
            words[index] =
                ((bytes[byteOffset] ?? 0) << 24) |
                    ((bytes[byteOffset + 1] ?? 0) << 16) |
                    ((bytes[byteOffset + 2] ?? 0) << 8) |
                    (bytes[byteOffset + 3] ?? 0);
        }
        for (let index = 16; index < 80; index += 1) {
            words[index] = rotateLeft(words[index - 3] ^
                words[index - 8] ^
                words[index - 14] ^
                words[index - 16], 1);
        }
        let a = h0;
        let b = h1;
        let c = h2;
        let d = h3;
        let e = h4;
        for (let index = 0; index < 80; index += 1) {
            let f;
            let k;
            if (index < 20) {
                f = (b & c) | (~b & d);
                k = 0x5a827999;
            }
            else if (index < 40) {
                f = b ^ c ^ d;
                k = 0x6ed9eba1;
            }
            else if (index < 60) {
                f = (b & c) | (b & d) | (c & d);
                k = 0x8f1bbcdc;
            }
            else {
                f = b ^ c ^ d;
                k = 0xca62c1d6;
            }
            const temp = (rotateLeft(a, 5) + f + e + k + (words[index] ?? 0)) >>> 0;
            e = d;
            d = c;
            c = rotateLeft(b, 30);
            b = a;
            a = temp;
        }
        h0 = (h0 + a) >>> 0;
        h1 = (h1 + b) >>> 0;
        h2 = (h2 + c) >>> 0;
        h3 = (h3 + d) >>> 0;
        h4 = (h4 + e) >>> 0;
    }
    return [h0, h1, h2, h3, h4]
        .map((word) => word.toString(16).padStart(8, '0'))
        .join('');
}
function shortenToolName(baseName, seed, forceHash = false) {
    if (!forceHash && baseName.length <= MCP_FUNCTION_TOOL_NAME_MAX_LENGTH) {
        return baseName;
    }
    const hashSuffix = getSha1Hex(seed).slice(0, MCP_FUNCTION_TOOL_HASH_LENGTH);
    const suffix = `_${hashSuffix}`;
    const stemLength = MCP_FUNCTION_TOOL_NAME_MAX_LENGTH - suffix.length;
    const stem = baseName.slice(0, stemLength).replace(/[_-]+$/g, '') || 'mcp';
    return `${stem}${suffix}`;
}
function buildPrefixedToolBaseName(serverName, toolName) {
    const serverPart = getSafeToolNamePart(serverName, 'server');
    const toolPart = getSafeToolNamePart(toolName, 'tool');
    return `mcp_${serverPart}__${toolPart}`;
}
function buildPrefixedToolNameOverrides(serverToolBatches, reservedNames) {
    const baseNameCounts = new Map();
    for (const { server, mcpTools } of serverToolBatches) {
        for (const mcpTool of mcpTools) {
            const baseName = buildPrefixedToolBaseName(server.name, mcpTool.name);
            baseNameCounts.set(baseName, (baseNameCounts.get(baseName) ?? 0) + 1);
        }
    }
    const candidates = [];
    for (const { server, serverIndex, mcpTools } of serverToolBatches) {
        mcpTools.forEach((mcpTool, toolIndex) => {
            const baseName = buildPrefixedToolBaseName(server.name, mcpTool.name);
            const seed = `${server.name}\0${mcpTool.name}`;
            const forceHash = (baseNameCounts.get(baseName) ?? 0) > 1 || reservedNames.has(baseName);
            candidates.push({
                batchKey: getToolNameOverrideKey(serverIndex, toolIndex),
                baseName,
                seed,
                initialName: shortenToolName(baseName, seed, forceHash),
                serverIndex,
                toolIndex,
            });
        });
    }
    const usedNames = new Set(reservedNames);
    const overrides = new Map();
    for (const candidate of candidates.sort((left, right) => {
        return (left.initialName.localeCompare(right.initialName) ||
            left.seed.localeCompare(right.seed) ||
            left.serverIndex - right.serverIndex ||
            left.toolIndex - right.toolIndex);
    })) {
        let publicName = candidate.initialName;
        let collisionIndex = 1;
        while (usedNames.has(publicName)) {
            publicName = shortenToolName(candidate.baseName, `${candidate.seed}\0${collisionIndex}`, true);
            collisionIndex += 1;
        }
        usedNames.add(publicName);
        overrides.set(candidate.batchKey, publicName);
    }
    return overrides;
}
async function resolveMcpToolMeta(server, runContext, toolName, args) {
    const resolver = server.toolMetaResolver;
    if (!resolver) {
        return undefined;
    }
    const context = {
        runContext,
        serverName: server.name,
        toolName,
        arguments: args,
    };
    const resolved = await resolver(context);
    if (resolved == null) {
        return undefined;
    }
    if (typeof resolved !== 'object' || Array.isArray(resolved)) {
        throw new TypeError('MCP tool meta resolver must return an object or null.');
    }
    return resolved;
}
function mcpToFunctionTool(mcpTool, server, convertSchemasToStrict, options = {}) {
    const toolName = options.toolNameOverride ?? mcpTool.name;
    const customDataByCall = new WeakMap();
    const serverErrorFunction = server.errorFunction;
    const mcpErrorFunction = serverErrorFunction !== undefined
        ? serverErrorFunction
        : options.errorFunction;
    const errorFunction = typeof mcpErrorFunction === 'function'
        ? (context, error) => mcpErrorFunction({ context, error })
        : mcpErrorFunction;
    async function invoke(input, runContext, details) {
        let args = {};
        if (typeof input === 'string' && input) {
            args = JSON.parse(input);
        }
        else if (typeof input === 'object' && input != null) {
            args = input;
        }
        const currentSpan = (0, tracing_1.getCurrentSpan)();
        if (currentSpan) {
            currentSpan.spanData['mcp_data'] = { server: server.name };
        }
        const meta = runContext
            ? await resolveMcpToolMeta(server, runContext, mcpTool.name, args)
            : undefined;
        const result = (server.useStructuredContent === true ||
            server.customDataExtractor !== undefined) &&
            server.callToolResult
            ? meta === undefined
                ? await server.callToolResult(mcpTool.name, args)
                : await server.callToolResult(mcpTool.name, args, meta)
            : {
                content: meta === undefined
                    ? await server.callTool(mcpTool.name, args)
                    : await server.callTool(mcpTool.name, args, meta),
            };
        const content = result.content;
        const resultMeta = result._meta ?? content._meta;
        const structuredContent = result.structuredContent ?? content.structuredContent;
        const isError = result.isError ?? content.isError;
        const toolOutput = server.useStructuredContent === true &&
            isError !== true &&
            structuredContent !== undefined
            ? JSON.stringify(structuredContent)
            : content.length === 1
                ? content[0]
                : content;
        if (runContext && details?.toolCall?.callId && server.customDataExtractor) {
            let byCall = customDataByCall.get(runContext);
            if (!byCall) {
                byCall = new Map();
                customDataByCall.set(runContext, byCall);
            }
            byCall.set(details.toolCall.callId, {
                runContext,
                serverName: server.name,
                toolName: mcpTool.name,
                toolDisplayName: toolName,
                arguments: cloneMcpCustomDataContextValue(args),
                resultMeta: cloneMcpCustomDataContextValue(resultMeta),
                structuredContent: cloneMcpCustomDataContextValue(structuredContent),
                isError,
                toolOutput: cloneMcpCustomDataContextValue(toolOutput),
            });
        }
        return toolOutput;
    }
    const schema = {
        ...mcpTool.inputSchema,
        type: mcpTool.inputSchema?.type ?? 'object',
        properties: mcpTool.inputSchema?.properties ?? {},
        required: mcpTool.inputSchema?.required ?? [],
        additionalProperties: mcpTool.inputSchema?.additionalProperties ?? false,
    };
    if (convertSchemasToStrict || schema.additionalProperties === true) {
        try {
            const strictSchema = ensureStrictJsonSchema(schema);
            return (0, tool_1.tool)({
                name: toolName,
                description: mcpTool.description || '',
                parameters: strictSchema,
                strict: true,
                execute: invoke,
                errorFunction,
                customDataExtractor: async (context) => {
                    const mcpContext = getMcpCustomDataContext(customDataByCall, context.runContext, context.toolCall.callId);
                    return mcpContext
                        ? (0, customData_1.maybeExtractToolOutputCustomData)(server.customDataExtractor, mcpContext)
                        : undefined;
                },
            });
        }
        catch (e) {
            logger_1.logger.warn(`Error converting MCP schema to strict mode: ${e}`);
        }
    }
    const nonStrictSchema = {
        ...schema,
        additionalProperties: true,
    };
    return (0, tool_1.tool)({
        name: toolName,
        description: mcpTool.description || '',
        parameters: nonStrictSchema,
        strict: false,
        execute: invoke,
        errorFunction,
        customDataExtractor: async (context) => {
            const mcpContext = getMcpCustomDataContext(customDataByCall, context.runContext, context.toolCall.callId);
            return mcpContext
                ? (0, customData_1.maybeExtractToolOutputCustomData)(server.customDataExtractor, mcpContext)
                : undefined;
        },
    });
}
function getMcpCustomDataContext(contexts, runContext, callId) {
    const byCall = contexts.get(runContext);
    const context = byCall?.get(callId);
    byCall?.delete(callId);
    return context;
}
function cloneMcpCustomDataContextValue(value) {
    try {
        return structuredClone(value);
    }
    catch {
        return value;
    }
}
/**
 * Ensures the given JSON schema is strict (no additional properties, required fields set).
 */
function ensureStrictJsonSchema(schema) {
    const out = {
        ...schema,
        additionalProperties: false,
    };
    if (!out.required)
        out.required = [];
    return out;
}
function attachCallToolResultMetadata(content, metadata) {
    const result = content;
    for (const [key, value] of Object.entries(metadata)) {
        if (typeof value === 'undefined') {
            continue;
        }
        Object.defineProperty(result, key, {
            value,
            enumerable: false,
            configurable: true,
        });
    }
    return result;
}
