// @openai/agents-core/dist/runState.mjs
import { z } from 'zod';
import { Agent } from "./agent.mjs";
import { getAgentToolSourceAgent } from "./agentToolSourceRegistry.mjs";
import { RunMessageOutputItem, RunToolApprovalItem, RunToolCallItem, RunToolCallOutputItem, RunToolSearchCallItem, RunToolSearchOutputItem, RunReasoningItem, RunHandoffCallItem, RunHandoffOutputItem, } from "./items.mjs";
import { RunContext } from "./runContext.mjs";
import { getTurnInput } from "./runner/items.mjs";
import { AgentToolUseTracker } from "./runner/toolUseTracker.mjs";
import { nextStepSchema } from "./runner/steps.mjs";
import { SystemError, UserError } from "./errors.mjs";
import { getGlobalTraceProvider } from "./tracing/provider.mjs";
import { Usage } from "./usage.mjs";
import { getCurrentTrace } from "./tracing/index.mjs";
import logger from "./logger.mjs";
import { handoff } from "./handoff.mjs";
import * as protocol from "./types/protocol.mjs";
import { SANDBOX_SESSION_STATE_VERSION } from "./sandbox/session.mjs";
import { safeExecute } from "./utils/safeExecute.mjs";
import { getClientToolSearchExecutor, getToolSearchRuntimeToolKey, } from "./tool.mjs";
import { getFunctionToolQualifiedName, toolQualifiedName, resolveFunctionToolCallName, } from "./toolIdentity.mjs";
import { getToolSearchExecution, getToolSearchOutputReplacementKey, resolveToolSearchCallId, } from "./utils/toolSearch.mjs";
import { executeCustomClientToolSearch, getClientToolSearchHelper, } from "./runner/toolSearch.mjs";
import { getSerializedApplyPatchToolPlaceholder, getSerializedComputerToolPlaceholder, getSerializedFunctionToolPlaceholder, getSerializedShellToolPlaceholder, } from "./sandbox/runtime/toolRehydration.mjs";
/**
 * The schema version of the serialized run state. This is used to ensure that the serialized
 * run state is compatible with the current version of the SDK.
 * If anything in this schema changes, the version will have to be incremented.
 *
 * Version history.
 * - 1.0: Initial serialized RunState schema.
 * - 1.1: Adds optional currentTurnInProgress, conversationId, and previousResponseId fields,
 *   plus broader tool_call_output_item rawItem variants for non-function tools. Older 1.0
 *   payloads remain readable but resumes may lack mid-turn or server-managed context precision.
 * - 1.2: Adds pendingAgentToolRuns for nested agent tool resumption.
 * - 1.3: Adds computer tool approval items to serialized tool_approval_item unions.
 * - 1.4: Adds optional toolInput to serialized run context.
 * - 1.5: Adds optional reasoningItemIdPolicy to preserve reasoning input policy across resume.
 * - 1.6: Adds optional requestId to serialized model responses.
 * - 1.7: Adds optional approval rejection messages.
 * - 1.8: Adds tool search item variants, batched computer actions, and GA computer tool
 *   aliasing to serialized run state payloads.
 * - 1.9: Adds optional sandbox session persistence with a versioned session-state
 *   envelope for sandbox-agent resume.
 * - 1.10: Adds optional stable agent identity keys so duplicate-name agent graphs can
 *   serialize and resume without ambiguous name resolution.
 * - 1.11: Allows null maxTurns to persist runs without a turn limit.
 * - 1.12: Adds optional missing function tool calls to processed responses.
 * - 1.13: Adds optional SDK-only customData on tool output run items.
 */
export const CURRENT_SCHEMA_VERSION = '1.13';
const SUPPORTED_SCHEMA_VERSIONS = [
    '1.0',
    '1.1',
    '1.2',
    '1.3',
    '1.4',
    '1.5',
    '1.6',
    '1.7',
    '1.8',
    '1.9',
    '1.10',
    '1.11',
    '1.12',
    CURRENT_SCHEMA_VERSION,
];
const $schemaVersion = z.enum(SUPPORTED_SCHEMA_VERSIONS);
const serializedAgentSchema = z.object({
    name: z.string(),
    identity: z.string().optional(),
});
const serializedSpanBase = z.object({
    object: z.literal('trace.span'),
    id: z.string(),
    trace_id: z.string(),
    parent_id: z.string().nullable(),
    started_at: z.string().nullable(),
    ended_at: z.string().nullable(),
    error: z
        .object({
        message: z.string(),
        data: z.record(z.string(), z.any()).optional(),
    })
        .nullable(),
    span_data: z.record(z.string(), z.any()),
});
const SerializedSpan = serializedSpanBase.extend({
    previous_span: z.lazy(() => SerializedSpan).optional(),
});
const requestUsageSchema = z.object({
    inputTokens: z.number(),
    outputTokens: z.number(),
    totalTokens: z.number(),
    inputTokensDetails: z.record(z.string(), z.number()).optional(),
    outputTokensDetails: z.record(z.string(), z.number()).optional(),
    endpoint: z.string().optional(),
});
const usageSchema = z.object({
    requests: z.number(),
    inputTokens: z.number(),
    outputTokens: z.number(),
    totalTokens: z.number(),
    inputTokensDetails: z.array(z.record(z.string(), z.number())).optional(),
    outputTokensDetails: z.array(z.record(z.string(), z.number())).optional(),
    requestUsageEntries: z.array(requestUsageSchema).optional(),
});
const modelResponseSchema = z.object({
    usage: usageSchema,
    output: z.array(protocol.OutputModelItem),
    responseId: z.string().optional(),
    requestId: z.string().optional(),
    providerData: z.record(z.string(), z.any()).optional(),
});
const itemSchema = z.discriminatedUnion('type', [
    z.object({
        type: z.literal('message_output_item'),
        rawItem: protocol.AssistantMessageItem,
        agent: serializedAgentSchema,
    }),
    z.object({
        type: z.literal('tool_search_call_item'),
        rawItem: protocol.ToolSearchCallItem,
        agent: serializedAgentSchema,
    }),
    z.object({
        type: z.literal('tool_search_output_item'),
        rawItem: protocol.ToolSearchOutputItem,
        agent: serializedAgentSchema,
    }),
    z.object({
        type: z.literal('tool_call_item'),
        rawItem: protocol.ToolCallItem.or(protocol.HostedToolCallItem),
        agent: serializedAgentSchema,
    }),
    z.object({
        type: z.literal('tool_call_output_item'),
        rawItem: protocol.FunctionCallResultItem.or(protocol.ComputerCallResultItem)
            .or(protocol.ShellCallResultItem)
            .or(protocol.ApplyPatchCallResultItem),
        agent: serializedAgentSchema,
        output: z.string(),
        customData: z.record(z.string(), z.any()).optional(),
    }),
    z.object({
        type: z.literal('reasoning_item'),
        rawItem: protocol.ReasoningItem,
        agent: serializedAgentSchema,
    }),
    z.object({
        type: z.literal('handoff_call_item'),
        rawItem: protocol.FunctionCallItem,
        agent: serializedAgentSchema,
    }),
    z.object({
        type: z.literal('handoff_output_item'),
        rawItem: protocol.FunctionCallResultItem,
        sourceAgent: serializedAgentSchema,
        targetAgent: serializedAgentSchema,
    }),
    z.object({
        type: z.literal('tool_approval_item'),
        rawItem: protocol.FunctionCallItem.or(protocol.HostedToolCallItem)
            .or(protocol.ComputerUseCallItem)
            .or(protocol.ShellCallItem)
            .or(protocol.ApplyPatchCallItem),
        agent: serializedAgentSchema,
        toolName: z.string().optional(),
    }),
]);
const serializedTraceSchema = z.object({
    object: z.literal('trace'),
    id: z.string(),
    workflow_name: z.string(),
    group_id: z.string().nullable(),
    metadata: z.record(z.string(), z.any()),
    // Populated only if the trace was created with a per-run tracingApiKey (e.g., Runner.run({ tracing: { apiKey } }))
    // and serialization opts in to include it. By default this is omitted to avoid persisting secrets.
    tracing_api_key: z.string().optional().nullable(),
});
const sandboxSessionStateEnvelopeSchema = z.object({
    version: z.literal(SANDBOX_SESSION_STATE_VERSION),
    backendId: z.string(),
    manifest: z.record(z.string(), z.any()),
    snapshot: z.record(z.string(), z.any()).nullable().optional(),
    snapshotFingerprint: z.string().nullable().optional(),
    snapshotFingerprintVersion: z.string().nullable().optional(),
    workspaceReady: z.boolean(),
    exposedPorts: z.record(z.string(), z.any()).optional(),
    providerState: z.record(z.string(), z.any()),
});
const sandboxSessionEntrySchema = z.object({
    backendId: z.string(),
    currentAgentKey: z.string(),
    currentAgentName: z.string(),
    sessionState: sandboxSessionStateEnvelopeSchema,
    preservedOwnedSession: z.boolean().optional(),
    reuseLiveSession: z.boolean().optional(),
});
const sandboxStateSchema = z.object({
    backendId: z.string(),
    currentAgentKey: z.string(),
    currentAgentName: z.string(),
    sessionState: sandboxSessionStateEnvelopeSchema,
    sessionsByAgent: z.record(z.string(), sandboxSessionEntrySchema),
});
const serializedProcessedResponseSchema = z.object({
    newItems: z.array(itemSchema),
    toolsUsed: z.array(z.string()),
    handoffs: z.array(z.object({
        toolCall: z.any(),
        handoff: z.any(),
    })),
    functions: z.array(z.object({
        toolCall: z.any(),
        tool: z.any(),
    })),
    functionToolsNotFound: z
        .array(z.object({
        toolCall: z.any(),
        toolName: z.string(),
    }))
        .optional(),
    computerActions: z.array(z.object({
        toolCall: z.any(),
        computer: z.any(),
    })),
    shellActions: z
        .array(z.object({
        toolCall: z.any(),
        shell: z.any(),
    }))
        .optional(),
    applyPatchActions: z
        .array(z.object({
        toolCall: z.any(),
        applyPatch: z.any(),
    }))
        .optional(),
    mcpApprovalRequests: z
        .array(z.object({
        requestItem: z.object({
            // protocol.HostedToolCallItem
            rawItem: z.object({
                type: z.literal('hosted_tool_call'),
                name: z.string(),
                arguments: z.string().optional(),
                status: z.string().optional(),
                output: z.string().optional(),
                // this always exists but marked as optional for early version compatibility; when releasing 1.0, we can remove the nullable and optional
                providerData: z.record(z.string(), z.any()).nullable().optional(),
            }),
        }),
        // HostedMCPTool
        mcpTool: z.object({
            type: z.literal('hosted_tool'),
            name: z.literal('hosted_mcp'),
            providerData: z.record(z.string(), z.any()),
        }),
    }))
        .optional(),
});
const guardrailFunctionOutputSchema = z.object({
    tripwireTriggered: z.boolean(),
    outputInfo: z.any(),
});
const toolGuardrailBehaviorSchema = z.discriminatedUnion('type', [
    z.object({ type: z.literal('allow') }),
    z.object({
        type: z.literal('rejectContent'),
        message: z.string(),
    }),
    z.object({ type: z.literal('throwException') }),
]);
const toolGuardrailFunctionOutputSchema = z.object({
    outputInfo: z.any().optional(),
    behavior: toolGuardrailBehaviorSchema,
});
const toolGuardrailMetadataSchema = z.object({
    type: z.union([z.literal('tool_input'), z.literal('tool_output')]),
    name: z.string(),
});
const inputGuardrailResultSchema = z.object({
    guardrail: z.object({
        type: z.literal('input'),
        name: z.string(),
    }),
    output: guardrailFunctionOutputSchema,
});
const outputGuardrailResultSchema = z.object({
    guardrail: z.object({
        type: z.literal('output'),
        name: z.string(),
    }),
    agentOutput: z.any(),
    agent: serializedAgentSchema,
    output: guardrailFunctionOutputSchema,
});
const toolInputGuardrailResultSchema = z.object({
    guardrail: toolGuardrailMetadataSchema.extend({
        type: z.literal('tool_input'),
    }),
    output: toolGuardrailFunctionOutputSchema,
});
const toolOutputGuardrailResultSchema = z.object({
    guardrail: toolGuardrailMetadataSchema.extend({
        type: z.literal('tool_output'),
    }),
    output: toolGuardrailFunctionOutputSchema,
});
export const SerializedRunState = z.object({
    $schemaVersion,
    currentTurn: z.number(),
    currentAgent: serializedAgentSchema,
    originalInput: z.string().or(z.array(protocol.ModelItem)),
    modelResponses: z.array(modelResponseSchema),
    context: z.object({
        usage: usageSchema,
        approvals: z.record(z.string(), z.object({
            approved: z.array(z.string()).or(z.boolean()),
            rejected: z.array(z.string()).or(z.boolean()),
            messages: z.record(z.string(), z.string()).optional(),
            stickyRejectMessage: z.string().optional(),
        })),
        context: z.record(z.string(), z.any()),
        toolInput: z.any().optional(),
    }),
    toolUseTracker: z.record(z.string(), z.array(z.string())),
    maxTurns: z.number().nullable(),
    currentAgentSpan: SerializedSpan.nullable().optional(),
    noActiveAgentRun: z.boolean(),
    inputGuardrailResults: z.array(inputGuardrailResultSchema),
    outputGuardrailResults: z.array(outputGuardrailResultSchema),
    toolInputGuardrailResults: z
        .array(toolInputGuardrailResultSchema)
        .optional()
        .default([]),
    toolOutputGuardrailResults: z
        .array(toolOutputGuardrailResultSchema)
        .optional()
        .default([]),
    currentTurnInProgress: z.boolean().optional(),
    currentStep: nextStepSchema.optional(),
    lastModelResponse: modelResponseSchema.optional(),
    generatedItems: z.array(itemSchema),
    pendingAgentToolRuns: z.record(z.string(), z.string()).optional().default({}),
    lastProcessedResponse: serializedProcessedResponseSchema.optional(),
    currentTurnPersistedItemCount: z.number().int().min(0).optional(),
    conversationId: z.string().optional(),
    previousResponseId: z.string().optional(),
    reasoningItemIdPolicy: z.enum(['preserve', 'omit']).optional(),
    trace: serializedTraceSchema.nullable(),
    sandbox: sandboxStateSchema.optional(),
});
/**
 * Serializable snapshot of an agent's run, including context, usage and trace.
 * While this class has publicly writable properties (prefixed with `_`), they are not meant to be
 * used directly. To read these properties, use the `RunResult` instead.
 *
 * Manipulation of the state directly can lead to unexpected behavior and should be avoided.
 * Instead, use the `approve` and `reject` methods to interact with the state.
 */
export class RunState {
    /**
     * Current turn number in the conversation.
     */
    _currentTurn = 0;
    /**
     * Whether the current turn has already been counted (useful when resuming mid-turn).
     */
    _currentTurnInProgress = false;
    /**
     * The agent currently handling the conversation.
     */
    _currentAgent;
    /**
     * The root agent that started the run.
     */
    #startingAgent;
    /**
     * Original user input prior to any processing.
     */
    _originalInput;
    /**
     * Responses from the model so far.
     */
    _modelResponses;
    /**
     * Conversation identifier when the server manages conversation history.
     */
    _conversationId;
    /**
     * Latest response identifier returned by the server for server-managed conversations.
     */
    _previousResponseId;
    /**
     * Runtime options that control how run items are converted into model turn input.
     * This value is serialized so resumed runs keep the same turn-input behavior.
     */
    _reasoningItemIdPolicy;
    /**
     * Effective model settings used for the most recent model call.
     */
    _lastModelSettings;
    /**
     * Active tracing span for the current agent if tracing is enabled.
     */
    _currentAgentSpan;
    /**
     * Run context tracking approvals, usage, and other metadata.
     */
    _context;
    /**
     * Runtime-only metadata for the current nested agent-tool invocation.
     */
    _agentToolInvocation;
    /**
     * The usage aggregated for this run. This includes per-request breakdowns when available.
     */
    get usage() {
        return this._context.usage;
    }
    /**
     * Tracks what tools each agent has used.
     */
    _toolUseTracker;
    /**
     * Serialized pending nested agent runs keyed by tool name and call id.
     */
    _pendingAgentToolRuns;
    /**
     * Items generated by the agent during the run.
     */
    _generatedItems;
    /**
     * Number of `_generatedItems` already flushed to session storage for the current turn.
     *
     * Persisting the entire turn on every save would duplicate responses and tool outputs.
     * Instead, `saveToSession` appends only the delta since the previous write. This counter
     * tracks how many generated run items from *this turn* were already written so the next
     * save can slice off only the new entries. When a turn is interrupted (e.g., awaiting tool
     * approval) and later resumed, we rewind the counter before continuing so the pending tool
     * output still gets stored.
     */
    _currentTurnPersistedItemCount;
    /**
     * Maximum allowed turns before forcing termination.
     */
    _maxTurns;
    /**
     * Whether the run has an active agent step in progress.
     */
    _noActiveAgentRun = true;
    /**
     * Last model response for the previous turn.
     */
    _lastTurnResponse;
    /**
     * Results from input guardrails applied to the run.
     */
    _inputGuardrailResults;
    /**
     * Results from output guardrails applied to the run.
     */
    _outputGuardrailResults;
    /**
     * Results from tool input guardrails applied during tool execution.
     */
    _toolInputGuardrailResults;
    /**
     * Results from tool output guardrails applied during tool execution.
     */
    _toolOutputGuardrailResults;
    /**
     * Next step computed for the agent to take.
     */
    _currentStep = undefined;
    /**
     * Indicates how the final output was produced for the current run.
     * This value is not serialized.
     */
    _finalOutputSource;
    /**
     * Parsed model response after applying guardrails and tools.
     */
    _lastProcessedResponse = undefined;
    /**
     * Trace associated with this run if tracing is enabled.
     */
    _trace = null;
    /**
     * Runtime-only tool_search-loaded tools, scoped by agent object and preserved across turns for
     * the lifetime of this in-memory run.
     */
    _toolSearchRuntimeToolsByAgent = new Map();
    /**
     * Persisted sandbox session metadata for sandbox-agent resume.
     */
    _sandbox = undefined;
    constructor(context, originalInput, startingAgent, maxTurns) {
        this._context = context;
        this._agentToolInvocation = undefined;
        this._originalInput = structuredClone(originalInput);
        this._modelResponses = [];
        this._currentAgentSpan = undefined;
        this._currentAgent = startingAgent;
        this.#startingAgent = startingAgent;
        this._reasoningItemIdPolicy = undefined;
        this._toolUseTracker = new AgentToolUseTracker();
        this._pendingAgentToolRuns = new Map();
        this._generatedItems = [];
        this._currentTurnPersistedItemCount = 0;
        this._maxTurns = maxTurns;
        this._inputGuardrailResults = [];
        this._outputGuardrailResults = [];
        this._toolInputGuardrailResults = [];
        this._toolOutputGuardrailResults = [];
        this._trace = getCurrentTrace();
    }
    /**
     * Updates server-managed conversation identifiers as a single operation.
     */
    setConversationContext(conversationId, previousResponseId) {
        this._conversationId = conversationId;
        this._previousResponseId = previousResponseId;
    }
    /**
     * Updates runtime options for converting run items into turn input.
     */
    setReasoningItemIdPolicy(policy) {
        this._reasoningItemIdPolicy = policy;
    }
    /**
     * Updates the agent span associated with the current run.
     */
    setCurrentAgentSpan(span) {
        this._currentAgentSpan = span;
    }
    /**
     * Clears the restored trace and current agent span from this run state.
     *
     * Use this before resuming a serialized state when the resumed run should attach
     * to the current ambient trace instead of the trace persisted in the state.
     */
    clearTrace() {
        this._trace = null;
        this._currentAgentSpan = undefined;
    }
    getOrCreateToolSearchRuntimeToolState(agent) {
        let state = this._toolSearchRuntimeToolsByAgent.get(agent);
        if (!state) {
            state = {
                anonymousEntries: [],
                keyedEntries: new Map(),
                nextOrder: 0,
            };
            this._toolSearchRuntimeToolsByAgent.set(agent, state);
        }
        return state;
    }
    recordToolSearchRuntimeTools(agent, toolSearchOutput, tools) {
        const runtimeState = this.getOrCreateToolSearchRuntimeToolState(agent);
        const entry = {
            order: runtimeState.nextOrder++,
            tools,
        };
        const replacementKey = getToolSearchOutputReplacementKey(toolSearchOutput);
        if (replacementKey) {
            runtimeState.keyedEntries.set(replacementKey, entry);
            return;
        }
        runtimeState.anonymousEntries.push(entry);
    }
    getToolSearchRuntimeTools(agent) {
        const runtimeState = this._toolSearchRuntimeToolsByAgent.get(agent);
        if (!runtimeState) {
            return [];
        }
        const dedupedTools = new Map();
        const orderedEntries = [
            ...runtimeState.keyedEntries.values(),
            ...runtimeState.anonymousEntries,
        ].sort((a, b) => a.order - b.order);
        let anonymousCounter = 0;
        for (const entry of orderedEntries) {
            for (const tool of entry.tools) {
                const key = getToolSearchRuntimeToolKey(tool) ??
                    `anonymous:${entry.order}:${anonymousCounter++}`;
                dedupedTools.set(key, tool);
            }
        }
        return [...dedupedTools.values()];
    }
    /**
     * Switches the active agent handling the run.
     */
    setCurrentAgent(agent) {
        this._currentAgent = agent;
    }
    /**
     * Returns the agent currently handling the run.
     */
    get currentAgent() {
        return this._currentAgent;
    }
    /**
     * Resets the counter that tracks how many items were persisted for the current turn.
     */
    resetTurnPersistence() {
        this._currentTurnPersistedItemCount = 0;
    }
    /**
     * Rewinds the persisted item counter when pending approvals require re-writing outputs.
     */
    rewindTurnPersistence(count) {
        if (count <= 0) {
            return;
        }
        this._currentTurnPersistedItemCount = Math.max(0, this._currentTurnPersistedItemCount - count);
    }
    /**
     * The history of the agent run. This includes the input items and the new items generated during the run.
     *
     * This can be used as inputs for the next agent run.
     */
    get history() {
        return getTurnInput(this._originalInput, this._generatedItems, this._reasoningItemIdPolicy);
    }
    /**
     * Returns all interruptions if the current step is an interruption otherwise returns an empty array.
     */
    getInterruptions() {
        if (this._currentStep?.type !== 'next_step_interruption') {
            return [];
        }
        const interruptions = this._currentStep.data.interruptions;
        return Array.isArray(interruptions)
            ? interruptions
            : [];
    }
    getPendingAgentToolRunKey(toolName, callId) {
        return `${toolName}:${callId}`;
    }
    getPendingAgentToolRun(toolName, callId) {
        return this._pendingAgentToolRuns.get(this.getPendingAgentToolRunKey(toolName, callId));
    }
    hasPendingAgentToolRun(toolName, callId) {
        return this._pendingAgentToolRuns.has(this.getPendingAgentToolRunKey(toolName, callId));
    }
    setPendingAgentToolRun(toolName, callId, serializedState) {
        this._pendingAgentToolRuns.set(this.getPendingAgentToolRunKey(toolName, callId), serializedState);
    }
    clearPendingAgentToolRun(toolName, callId) {
        this._pendingAgentToolRuns.delete(this.getPendingAgentToolRunKey(toolName, callId));
    }
    /**
     * Approves a tool call requested by the agent through an interruption and approval item request.
     *
     * To approve the request use this method and then run the agent again with the same state object
     * to continue the execution.
     *
     * By default it will only approve the current tool call. To allow the tool to be used multiple
     * times throughout the run, set the `alwaysApprove` option to `true`.
     *
     * @param approvalItem - The tool call approval item to approve.
     * @param options - Options for the approval.
     * @param options.alwaysApprove - Approve this tool for all future calls in this run.
     */
    approve(approvalItem, options = {
        alwaysApprove: false,
    }) {
        this._context.approveTool(approvalItem, options);
    }
    /**
     * Rejects a tool call requested by the agent through an interruption and approval item request.
     *
     * To reject the request use this method and then run the agent again with the same state object
     * to continue the execution.
     *
     * By default it will only reject the current tool call. To reject the tool for all future
     * calls throughout the run, set the `alwaysReject` option to `true`.
     *
     * When `message` is provided, it is used as the rejection text sent to the model.
     * Otherwise, `toolErrorFormatter` (if configured) or the SDK default is used.
     *
     * @param approvalItem - The tool call approval item to reject.
     * @param options - Options for the rejection.
     * @param options.alwaysReject - Reject this tool for all future calls in this run.
     * @param options.message - The rejection text sent to the model.
     *   If not provided, `toolErrorFormatter` (if configured) or the SDK default is used.
     */
    reject(approvalItem, options = {
        alwaysReject: false,
    }) {
        this._context.rejectTool(approvalItem, options);
    }
    /**
     * Serializes the run state to a JSON object.
     *
     * This method is used to serialize the run state to a JSON object that can be used to
     * resume the run later.
     *
     * @returns The serialized run state.
     */
    /**
     * Serializes the run state. By default, tracing API keys are omitted to prevent
     * accidental persistence of secrets. Pass `includeTracingApiKey: true` only when you
     * intentionally need to migrate a run along with its tracing credentials (e.g., to
     * rehydrate in a separate process that lacks the original environment variables).
     */
    toJSON(options = {}) {
        const agentIdentity = buildAgentIdentityMap(this.#startingAgent);
        const includeTracingApiKey = options.includeTracingApiKey === true;
        const contextJson = this._context.toJSON();
        const output = {
            $schemaVersion: CURRENT_SCHEMA_VERSION,
            currentTurn: this._currentTurn,
            currentAgent: serializeAgentReference(this._currentAgent, agentIdentity.byAgent),
            originalInput: this._originalInput,
            modelResponses: this._modelResponses.map((response) => {
                return {
                    usage: {
                        requests: response.usage.requests,
                        inputTokens: response.usage.inputTokens,
                        outputTokens: response.usage.outputTokens,
                        totalTokens: response.usage.totalTokens,
                        inputTokensDetails: response.usage.inputTokensDetails,
                        outputTokensDetails: response.usage.outputTokensDetails,
                        ...(response.usage.requestUsageEntries &&
                            response.usage.requestUsageEntries.length > 0
                            ? {
                                requestUsageEntries: response.usage.requestUsageEntries.map((entry) => ({
                                    inputTokens: entry.inputTokens,
                                    outputTokens: entry.outputTokens,
                                    totalTokens: entry.totalTokens,
                                    inputTokensDetails: entry.inputTokensDetails,
                                    outputTokensDetails: entry.outputTokensDetails,
                                    ...(entry.endpoint ? { endpoint: entry.endpoint } : {}),
                                })),
                            }
                            : {}),
                    },
                    output: response.output,
                    responseId: response.responseId,
                    requestId: response.requestId,
                    providerData: response.providerData,
                };
            }),
            context: contextJson,
            toolUseTracker: this._toolUseTracker.toJSON({
                agentIdentityKeys: agentIdentity.byAgent,
            }),
            maxTurns: this._maxTurns,
            currentAgentSpan: this._currentAgentSpan?.toJSON(),
            noActiveAgentRun: this._noActiveAgentRun,
            currentTurnInProgress: this._currentTurnInProgress,
            inputGuardrailResults: this._inputGuardrailResults,
            outputGuardrailResults: this._outputGuardrailResults.map((r) => ({
                ...r,
                agent: serializeAgentReference(r.agent, agentIdentity.byAgent),
            })),
            toolInputGuardrailResults: this._toolInputGuardrailResults,
            toolOutputGuardrailResults: this._toolOutputGuardrailResults,
            currentStep: serializeCurrentStep(this._currentStep, agentIdentity.byAgent),
            lastModelResponse: this._lastTurnResponse,
            generatedItems: this._generatedItems.map((item) => serializeRunItem(item, agentIdentity.byAgent)),
            pendingAgentToolRuns: Object.fromEntries(this._pendingAgentToolRuns.entries()),
            currentTurnPersistedItemCount: this._currentTurnPersistedItemCount,
            lastProcessedResponse: this._lastProcessedResponse
                ? serializeProcessedResponse(this._lastProcessedResponse, agentIdentity.byAgent)
                : undefined,
            conversationId: this._conversationId,
            previousResponseId: this._previousResponseId,
            reasoningItemIdPolicy: this._reasoningItemIdPolicy,
            trace: this._trace
                ? this._trace.toJSON({ includeTracingApiKey })
                : null,
            sandbox: this._sandbox,
        };
        // parsing the schema to ensure the output is valid for reparsing
        const parsed = SerializedRunState.safeParse(output);
        if (!parsed.success) {
            throw new SystemError(`Failed to serialize run state. ${parsed.error.message}`);
        }
        return parsed.data;
    }
    /**
     * Serializes the run state to a string.
     *
     * This method is used to serialize the run state to a string that can be used to
     * resume the run later.
     *
     * @returns The serialized run state.
     */
    toString(options = {}) {
        return JSON.stringify(this.toJSON(options));
    }
    /**
     * Deserializes a run state from a string.
     *
     * This method is used to deserialize a run state from a string that was serialized using the
     * `toString` method.
     */
    static async fromString(initialAgent, str) {
        return buildRunStateFromString(initialAgent, str);
    }
    static async fromStringWithContext(initialAgent, str, context, options = {}) {
        return buildRunStateFromString(initialAgent, str, {
            contextOverride: context,
            contextStrategy: options.contextStrategy,
        });
    }
}
async function buildRunStateFromString(initialAgent, str, options = {}) {
    const [parsingError, jsonResult] = await safeExecute(() => JSON.parse(str));
    if (parsingError) {
        throw new UserError(`Failed to parse run state. ${parsingError instanceof Error ? parsingError.message : String(parsingError)}`);
    }
    const currentSchemaVersion = jsonResult.$schemaVersion;
    if (!currentSchemaVersion) {
        throw new UserError('Run state is missing schema version');
    }
    if (!SUPPORTED_SCHEMA_VERSIONS.includes(currentSchemaVersion)) {
        throw new UserError(`Run state schema version ${currentSchemaVersion} is not supported. Please use version ${CURRENT_SCHEMA_VERSION}.`);
    }
    const stateJson = SerializedRunState.parse(jsonResult);
    assertSchemaVersionSupportsToolSearch(currentSchemaVersion, stateJson);
    assertSchemaVersionSupportsCustomData(currentSchemaVersion, stateJson);
    return buildRunStateFromJson(initialAgent, stateJson, options);
}
function assertSchemaVersionSupportsToolSearch(schemaVersion, stateJson) {
    if (schemaVersion === '1.8' ||
        schemaVersion === '1.9' ||
        schemaVersion === '1.10' ||
        schemaVersion === '1.11' ||
        schemaVersion === '1.12' ||
        schemaVersion === CURRENT_SCHEMA_VERSION) {
        return;
    }
    if (!containsSerializedToolSearchState(stateJson)) {
        return;
    }
    throw new UserError(`Run state schema version ${schemaVersion} does not support tool_search items. Please reserialize the run state with schema ${CURRENT_SCHEMA_VERSION}.`);
}
function assertSchemaVersionSupportsCustomData(schemaVersion, stateJson) {
    if (schemaVersion === CURRENT_SCHEMA_VERSION) {
        return;
    }
    if (!containsSerializedToolOutputCustomData(stateJson)) {
        return;
    }
    throw new UserError(`Run state schema version ${schemaVersion} does not support tool output customData. Please reserialize the run state with schema ${CURRENT_SCHEMA_VERSION}.`);
}
function schemaVersionSupportsAgentIdentity(schemaVersion) {
    return (schemaVersion === '1.10' ||
        schemaVersion === '1.11' ||
        schemaVersion === '1.12' ||
        schemaVersion === CURRENT_SCHEMA_VERSION);
}
function containsSerializedToolSearchState(stateJson) {
    return (containsToolSearchProtocolItems(stateJson.originalInput) ||
        containsToolSearchInModelResponses(stateJson.modelResponses) ||
        containsToolSearchInModelResponse(stateJson.lastModelResponse) ||
        containsToolSearchRunItems(stateJson.generatedItems) ||
        containsToolSearchInProcessedResponse(stateJson.lastProcessedResponse));
}
function containsToolSearchInModelResponses(modelResponses) {
    return modelResponses.some(containsToolSearchInModelResponse);
}
function containsToolSearchInModelResponse(modelResponse) {
    return Boolean(modelResponse?.output.some((item) => isToolSearchProtocolType(item.type)));
}
function containsToolSearchRunItems(items) {
    return Boolean(items?.some((item) => item.type === 'tool_search_call_item' ||
        item.type === 'tool_search_output_item' ||
        isToolSearchProtocolType(item.rawItem?.type)));
}
function containsToolSearchProtocolItems(items) {
    return Array.isArray(items)
        ? items.some((item) => isToolSearchProtocolType(item.type))
        : false;
}
function containsToolSearchInProcessedResponse(processedResponse) {
    return containsToolSearchRunItems(processedResponse?.newItems);
}
function containsSerializedToolOutputCustomData(stateJson) {
    return (containsToolOutputCustomDataRunItems(stateJson.generatedItems) ||
        containsToolOutputCustomDataRunItems(stateJson.lastProcessedResponse?.newItems));
}
function containsToolOutputCustomDataRunItems(items) {
    return Boolean(items?.some((item) => item.type === 'tool_call_output_item' &&
        Object.prototype.hasOwnProperty.call(item, 'customData')));
}
function isToolSearchProtocolType(type) {
    return type === 'tool_search_call' || type === 'tool_search_output';
}
function collectSerializedRuntimeToolKeys(value, runtimeToolKeys, namespace) {
    if (!value || typeof value !== 'object') {
        return;
    }
    const candidate = value;
    if (candidate.type === 'namespace' && Array.isArray(candidate.tools)) {
        const nestedNamespace = typeof candidate.name === 'string' && candidate.name.length > 0
            ? candidate.name
            : namespace;
        for (const nestedTool of candidate.tools) {
            collectSerializedRuntimeToolKeys(nestedTool, runtimeToolKeys, nestedNamespace);
        }
        return;
    }
    const explicitNamespace = typeof candidate.namespace === 'string' && candidate.namespace.length > 0
        ? candidate.namespace
        : namespace;
    if (candidate.type === 'function') {
        if (typeof candidate.name !== 'string' || candidate.name.length === 0) {
            return;
        }
        runtimeToolKeys.add(toolQualifiedName(candidate.name, explicitNamespace) ?? candidate.name);
        return;
    }
    if (candidate.type === 'mcp' &&
        typeof candidate.server_label === 'string' &&
        candidate.server_label.length > 0) {
        runtimeToolKeys.add(`mcp:${candidate.server_label}`);
        return;
    }
    if (!candidate.providerData || typeof candidate.providerData !== 'object') {
        return;
    }
    collectSerializedRuntimeToolKeys(candidate.providerData, runtimeToolKeys, explicitNamespace);
}
function getSerializedRuntimeToolKeys(toolSearchOutput) {
    const runtimeToolKeys = new Set();
    for (const tool of toolSearchOutput.tools) {
        collectSerializedRuntimeToolKeys(tool, runtimeToolKeys);
    }
    return runtimeToolKeys;
}
function getRuntimeToolKeys(runtimeTools, options = {}) {
    const runtimeToolKeys = new Set();
    for (const tool of runtimeTools) {
        const runtimeToolKey = getToolSearchRuntimeToolKey(tool);
        if (!runtimeToolKey) {
            if (options.allowUnsupported) {
                continue;
            }
            throw new UserError('Client tool_search execute() returned an unsupported runtime tool during RunState rehydration.');
        }
        runtimeToolKeys.add(runtimeToolKey);
    }
    return runtimeToolKeys;
}
function formatRuntimeToolKeys(runtimeToolKeys) {
    return [...runtimeToolKeys].sort().join(', ');
}
function assertRuntimeToolKeysMatch(args) {
    const { agent, toolSearchCall, expectedRuntimeToolKeys, runtimeTools } = args;
    if (expectedRuntimeToolKeys.size === 0) {
        return;
    }
    const actualRuntimeToolKeys = getRuntimeToolKeys(runtimeTools);
    const hasExpectedKeys = [...expectedRuntimeToolKeys].every((runtimeToolKey) => actualRuntimeToolKeys.has(runtimeToolKey));
    const hasActualKeys = [...actualRuntimeToolKeys].every((runtimeToolKey) => expectedRuntimeToolKeys.has(runtimeToolKey));
    if (hasExpectedKeys && hasActualKeys) {
        return;
    }
    const callId = resolveToolSearchCallId(toolSearchCall);
    throw new UserError(`RunState cannot resume custom client tool_search call ${callId} for agent ${agent.name} because the registered execute callback returned runtime tools [${formatRuntimeToolKeys(actualRuntimeToolKeys)}] but the serialized state expects [${formatRuntimeToolKeys(expectedRuntimeToolKeys)}].`);
}
async function getConfiguredAgentTools(args) {
    const { agent, context, configuredToolsByAgent } = args;
    const existing = configuredToolsByAgent.get(agent);
    if (existing) {
        return existing;
    }
    const configuredTools = (await agent.getAllTools(context));
    configuredToolsByAgent.set(agent, configuredTools);
    return configuredTools;
}
async function rehydrateToolSearchRuntimeTools(state) {
    const configuredToolsByAgent = new Map();
    const pendingToolSearchCalls = new Map();
    for (const item of state._generatedItems) {
        if (item instanceof RunToolSearchCallItem) {
            if (getToolSearchExecution(item.rawItem) === 'server') {
                continue;
            }
            const callId = resolveToolSearchCallId(item.rawItem);
            const agent = item.agent;
            const pendingCallsById = pendingToolSearchCalls.get(agent) ??
                new Map();
            pendingCallsById.set(callId, {
                agent: item.agent,
                toolSearchCall: item.rawItem,
            });
            pendingToolSearchCalls.set(agent, pendingCallsById);
            continue;
        }
        if (!(item instanceof RunToolSearchOutputItem)) {
            continue;
        }
        if (getToolSearchExecution(item.rawItem) === 'server') {
            continue;
        }
        const configuredTools = await getConfiguredAgentTools({
            agent: item.agent,
            context: state._context,
            configuredToolsByAgent,
        });
        const configuredToolKeys = getRuntimeToolKeys(configuredTools, {
            allowUnsupported: true,
        });
        const expectedRuntimeToolKeys = new Set([...getSerializedRuntimeToolKeys(item.rawItem)].filter((runtimeToolKey) => !configuredToolKeys.has(runtimeToolKey)));
        if (expectedRuntimeToolKeys.size === 0) {
            continue;
        }
        const callId = resolveToolSearchCallId(item.rawItem);
        const pendingCall = pendingToolSearchCalls
            .get(item.agent)
            ?.get(callId);
        if (!pendingCall) {
            throw new UserError(`RunState cannot resume custom client tool_search output ${callId} for agent ${item.agent.name} because the serialized state is missing the matching tool_search call item.`);
        }
        if (!pendingCall.runtimeTools) {
            const availableTools = [
                ...configuredTools,
                ...state.getToolSearchRuntimeTools(pendingCall.agent),
            ];
            const toolSearchTool = getClientToolSearchHelper(configuredTools);
            if (!toolSearchTool || !getClientToolSearchExecutor(toolSearchTool)) {
                throw new UserError(`RunState cannot resume custom client tool_search call ${callId} for agent ${pendingCall.agent.name} because the agent no longer provides toolSearchTool({ execution: "client", execute }).`);
            }
            const { runtimeTools } = await executeCustomClientToolSearch({
                agent: pendingCall.agent,
                runContext: state._context,
                toolSearchCall: pendingCall.toolSearchCall,
                toolSearchTool,
                tools: availableTools,
            });
            const rehydratedRuntimeTools = runtimeTools.filter((tool) => {
                const runtimeToolKey = getToolSearchRuntimeToolKey(tool);
                if (!runtimeToolKey) {
                    throw new UserError('Client tool_search execute() returned an unsupported runtime tool during RunState rehydration.');
                }
                return !configuredToolKeys.has(runtimeToolKey);
            });
            assertRuntimeToolKeysMatch({
                agent: pendingCall.agent,
                toolSearchCall: pendingCall.toolSearchCall,
                expectedRuntimeToolKeys,
                runtimeTools: rehydratedRuntimeTools,
            });
            pendingCall.runtimeTools = rehydratedRuntimeTools;
        }
        const runtimeTools = pendingCall.runtimeTools;
        if (!runtimeTools) {
            throw new UserError(`RunState cannot resume custom client tool_search call ${callId} for agent ${pendingCall.agent.name} because no runtime tools were rehydrated.`);
        }
        state.recordToolSearchRuntimeTools(pendingCall.agent, item.rawItem, runtimeTools);
    }
}
async function buildRunStateFromJson(initialAgent, stateJson, options = {}) {
    const agentMap = schemaVersionSupportsAgentIdentity(stateJson.$schemaVersion)
        ? buildAgentIdentityMap(initialAgent).byIdentity
        : buildAgentMap(initialAgent);
    const contextOverride = options.contextOverride;
    const contextStrategy = options.contextStrategy ?? 'merge';
    //
    // Rebuild the context
    //
    const context = contextOverride ??
        new RunContext(stateJson.context.context);
    if (contextOverride) {
        if (contextStrategy === 'merge') {
            context._mergeApprovals(stateJson.context.approvals);
        }
    }
    else {
        context._rebuildApprovals(stateJson.context.approvals);
    }
    const shouldRestoreToolInput = !contextOverride || contextStrategy === 'merge';
    if (shouldRestoreToolInput &&
        typeof stateJson.context.toolInput !== 'undefined' &&
        typeof context.toolInput === 'undefined') {
        context.toolInput = stateJson.context.toolInput;
    }
    //
    // Find the current agent from the initial agent
    //
    const currentAgent = resolveSerializedAgent(stateJson.currentAgent, agentMap);
    const state = new RunState(context, '', initialAgent, stateJson.maxTurns);
    state._currentAgent = currentAgent;
    state._currentTurn = stateJson.currentTurn;
    state._currentTurnInProgress = stateJson.currentTurnInProgress ?? false;
    state._conversationId = stateJson.conversationId ?? undefined;
    state._previousResponseId = stateJson.previousResponseId ?? undefined;
    state._reasoningItemIdPolicy = stateJson.reasoningItemIdPolicy ?? undefined;
    // rebuild tool use tracker
    state._toolUseTracker = new AgentToolUseTracker();
    for (const [agentName, toolNames] of Object.entries(stateJson.toolUseTracker)) {
        const agent = agentMap.get(agentName);
        if (!agent) {
            throw new UserError(`Agent ${agentName} not found`);
        }
        state._toolUseTracker.addToolUse(agent, toolNames, {
            allowEmpty: true,
        });
    }
    state._pendingAgentToolRuns = new Map(Object.entries(stateJson.pendingAgentToolRuns ?? {}));
    // rebuild current agent span
    if (stateJson.currentAgentSpan) {
        if (!stateJson.trace) {
            logger.warn('Trace is not set, skipping tracing setup');
        }
        const trace = getGlobalTraceProvider().createTrace({
            traceId: stateJson.trace?.id,
            name: stateJson.trace?.workflow_name,
            groupId: stateJson.trace?.group_id ?? undefined,
            metadata: stateJson.trace?.metadata,
            tracingApiKey: stateJson.trace?.tracing_api_key ?? undefined,
        });
        state._currentAgentSpan = deserializeSpan(trace, stateJson.currentAgentSpan);
        state._trace = trace;
    }
    state._noActiveAgentRun = stateJson.noActiveAgentRun;
    state._inputGuardrailResults =
        stateJson.inputGuardrailResults;
    state._outputGuardrailResults = stateJson.outputGuardrailResults.map((r) => ({
        ...r,
        agent: resolveSerializedAgent(r.agent, agentMap),
    }));
    state._toolInputGuardrailResults =
        stateJson.toolInputGuardrailResults;
    state._toolOutputGuardrailResults =
        stateJson.toolOutputGuardrailResults;
    state._currentStep = stateJson.currentStep;
    state._originalInput = stateJson.originalInput;
    state._modelResponses = stateJson.modelResponses.map(deserializeModelResponse);
    state._lastTurnResponse = stateJson.lastModelResponse
        ? deserializeModelResponse(stateJson.lastModelResponse)
        : undefined;
    state._generatedItems = stateJson.generatedItems.map((item) => deserializeItem(item, agentMap));
    state._currentTurnPersistedItemCount =
        stateJson.currentTurnPersistedItemCount ?? 0;
    state._sandbox = stateJson.sandbox ?? undefined;
    await rehydrateToolSearchRuntimeTools(state);
    state._lastProcessedResponse = stateJson.lastProcessedResponse
        ? await deserializeProcessedResponse(agentMap, state, stateJson.lastProcessedResponse)
        : undefined;
    if (stateJson.currentStep?.type === 'next_step_handoff') {
        state._currentStep = {
            type: 'next_step_handoff',
            newAgent: resolveSerializedAgent(stateJson.currentStep.newAgent, agentMap),
        };
    }
    else if (stateJson.currentStep?.type === 'next_step_interruption') {
        state._currentStep = {
            type: 'next_step_interruption',
            data: {
                ...stateJson.currentStep.data,
                interruptions: deserializeInterruptions(stateJson.currentStep.data?.interruptions, agentMap, state._currentAgent),
            },
        };
    }
    return state;
}
/**
 * @internal
 */
export async function rehydrateProcessedResponseTools(initialAgent, state, executionTools) {
    if (!state._lastProcessedResponse) {
        return;
    }
    const agentIdentity = buildAgentIdentityMap(initialAgent);
    const serializedProcessedResponse = serializeProcessedResponse(state._lastProcessedResponse, agentIdentity.byAgent);
    state._lastProcessedResponse = await deserializeProcessedResponse(agentIdentity.byIdentity, state, serializedProcessedResponse, {
        executionTools,
        allowSerializedExecutionToolPlaceholder: false,
    });
}
/**
 * @internal
 */
export function buildAgentMap(initialAgent) {
    const map = new Map();
    const visitedAgents = new Set();
    const queue = [initialAgent];
    while (queue.length > 0) {
        const currentAgent = queue.shift();
        if (visitedAgents.has(currentAgent)) {
            continue;
        }
        visitedAgents.add(currentAgent);
        const existingAgent = map.get(currentAgent.name);
        if (existingAgent && existingAgent !== currentAgent) {
            throw new UserError(`Duplicate agent name "${currentAgent.name}" detected. Use unique agent names when serializing RunState.`);
        }
        map.set(currentAgent.name, currentAgent);
        for (const handoff of currentAgent.handoffs) {
            if (handoff instanceof Agent) {
                queue.push(handoff);
            }
            else if (handoff.agent) {
                queue.push(handoff.agent);
            }
        }
        for (const tool of currentAgent.tools) {
            const sourceAgent = getAgentToolSourceAgent(tool);
            if (sourceAgent) {
                queue.push(sourceAgent);
            }
        }
    }
    return map;
}
/**
 * @internal
 */
export function buildAgentIdentityMap(initialAgent) {
    const agents = collectAgentGraph(initialAgent);
    const groups = new Map();
    const literalNames = new Set();
    for (const entry of agents) {
        literalNames.add(entry.agent.name);
        const group = groups.get(entry.agent.name) ?? [];
        group.push(entry);
        groups.set(entry.agent.name, group);
    }
    const byIdentity = new Map();
    const byAgent = new Map();
    const usedIdentities = new Set();
    for (const [agentName, group] of groups) {
        const sortedGroup = group.length === 1
            ? group
            : [...group].sort((left, right) => {
                if (left.agent === initialAgent) {
                    return -1;
                }
                if (right.agent === initialAgent) {
                    return 1;
                }
                const leftSignature = getAgentIdentitySignature(left.agent);
                const rightSignature = getAgentIdentitySignature(right.agent);
                if (leftSignature !== rightSignature) {
                    return leftSignature < rightSignature ? -1 : 1;
                }
                return left.index - right.index;
            });
        let nextSuffix = 0;
        for (const { agent } of sortedGroup) {
            let identity;
            do {
                identity =
                    nextSuffix === 0 ? agentName : `${agentName}#${nextSuffix + 1}`;
                nextSuffix += 1;
            } while (usedIdentities.has(identity) ||
                (identity !== agent.name && literalNames.has(identity)));
            usedIdentities.add(identity);
            byIdentity.set(identity, agent);
            byAgent.set(agent, identity);
        }
    }
    return { byIdentity, byAgent };
}
function collectAgentGraph(initialAgent) {
    const agents = [];
    const visitedAgents = new Set();
    const queue = [initialAgent];
    while (queue.length > 0) {
        const currentAgent = queue.shift();
        if (visitedAgents.has(currentAgent)) {
            continue;
        }
        visitedAgents.add(currentAgent);
        agents.push({ agent: currentAgent, index: agents.length });
        for (const handoff of currentAgent.handoffs) {
            if (handoff instanceof Agent) {
                queue.push(handoff);
            }
            else if (handoff.agent) {
                queue.push(handoff.agent);
            }
        }
        for (const tool of currentAgent.tools) {
            const sourceAgent = getAgentToolSourceAgent(tool);
            if (sourceAgent) {
                queue.push(sourceAgent);
            }
        }
    }
    return agents;
}
function getAgentIdentitySignature(agent) {
    const sandboxAgent = agent;
    const signature = {
        type: agent.constructor?.name,
        name: agent.name,
        handoffDescription: agent.handoffDescription,
        instructions: summarizeIdentityValue(agent.instructions),
        prompt: summarizeIdentityValue(agent.prompt),
        model: summarizeIdentityValue(agent.model),
        modelSettings: summarizeIdentityValue(agent.modelSettings),
        tools: agent.tools.map(summarizeToolIdentity),
        handoffs: agent.handoffs.map((entry) => entry instanceof Agent
            ? { type: 'agent', name: entry.name }
            : {
                type: 'handoff',
                toolName: entry.toolName,
                agentName: entry.agentName,
                targetName: entry.agent?.name,
            }),
        mcpServers: agent.mcpServers.map(summarizeIdentityValue),
        mcpConfig: summarizeIdentityValue(agent.mcpConfig),
        inputGuardrails: agent.inputGuardrails.map(summarizeIdentityValue),
        outputGuardrails: agent.outputGuardrails.map(summarizeIdentityValue),
        outputType: summarizeIdentityValue(agent.outputType),
        toolUseBehavior: summarizeIdentityValue(agent.toolUseBehavior),
        resetToolChoice: agent.resetToolChoice,
        defaultManifest: summarizeIdentityValue(sandboxAgent.defaultManifest),
        baseInstructions: summarizeIdentityValue(sandboxAgent.baseInstructions),
        capabilities: sandboxAgent.capabilities?.map(summarizeIdentityValue),
        runAs: summarizeIdentityValue(sandboxAgent.runAs),
    };
    return stableStringify(signature);
}
function summarizeToolIdentity(tool) {
    return {
        type: tool.type,
        name: tool.name,
        namespace: tool.namespace,
        strict: tool.strict,
        parameters: summarizeIdentityValue(tool.parameters),
    };
}
function summarizeIdentityValue(value) {
    return normalizeForIdentity(value, new WeakSet(), 0);
}
function normalizeForIdentity(value, seen, depth) {
    if (value === null || typeof value === 'undefined') {
        return value;
    }
    if (typeof value === 'string' ||
        typeof value === 'number' ||
        typeof value === 'boolean') {
        return value;
    }
    if (typeof value === 'function') {
        return `[function:${value.name || 'anonymous'}]`;
    }
    if (typeof value !== 'object') {
        return String(value);
    }
    if (seen.has(value)) {
        return '[circular]';
    }
    if (depth >= 4) {
        return `[${value.constructor?.name ?? 'Object'}]`;
    }
    seen.add(value);
    if (Array.isArray(value)) {
        return value.map((item) => normalizeForIdentity(item, seen, depth + 1));
    }
    if (value instanceof Map) {
        return [...value.entries()]
            .map(([key, entryValue]) => [
            normalizeForIdentity(key, seen, depth + 1),
            normalizeForIdentity(entryValue, seen, depth + 1),
        ])
            .sort((left, right) => stableStringify(left).localeCompare(stableStringify(right)));
    }
    if (value instanceof Set) {
        return [...value.values()]
            .map((entry) => normalizeForIdentity(entry, seen, depth + 1))
            .sort((left, right) => stableStringify(left).localeCompare(stableStringify(right)));
    }
    const record = value;
    const normalized = {
        constructor: value.constructor?.name,
    };
    for (const key of Object.keys(record).sort()) {
        normalized[key] = normalizeForIdentity(record[key], seen, depth + 1);
    }
    return normalized;
}
function stableStringify(value) {
    return JSON.stringify(value, (_key, currentValue) => {
        if (!currentValue ||
            typeof currentValue !== 'object' ||
            Array.isArray(currentValue)) {
            return currentValue;
        }
        return Object.fromEntries(Object.entries(currentValue).sort(([left], [right]) => left.localeCompare(right)));
    });
}
function serializeAgentReference(agent, agentIdentityKeys) {
    const identity = agentIdentityKeys.get(agent);
    if (!identity || identity === agent.name) {
        return { name: agent.name };
    }
    return { name: agent.name, identity };
}
function resolveSerializedAgent(serializedAgent, agentMap, fallbackAgent) {
    const identity = serializedAgent.identity ?? serializedAgent.name;
    const agent = agentMap.get(identity);
    if (agent) {
        return agent;
    }
    if (!serializedAgent.identity && fallbackAgent) {
        return fallbackAgent;
    }
    if (serializedAgent.identity) {
        throw new UserError(`Agent identity ${serializedAgent.identity} not found`);
    }
    throw new UserError(`Agent ${serializedAgent.name} not found`);
}
function serializeRunItem(item, agentIdentityKeys) {
    const serialized = item.toJSON();
    switch (item.type) {
        case 'handoff_output_item':
            serialized.sourceAgent = serializeAgentReference(item.sourceAgent, agentIdentityKeys);
            serialized.targetAgent = serializeAgentReference(item.targetAgent, agentIdentityKeys);
            return serialized;
        default:
            serialized.agent = serializeAgentReference(item.agent, agentIdentityKeys);
            return serialized;
    }
}
function serializeCurrentStep(currentStep, agentIdentityKeys) {
    if (!currentStep) {
        return undefined;
    }
    if (currentStep.type === 'next_step_handoff') {
        return {
            ...currentStep,
            newAgent: serializeAgentReference(currentStep.newAgent, agentIdentityKeys),
        };
    }
    if (currentStep.type === 'next_step_interruption') {
        const interruptions = Array.isArray(currentStep.data?.interruptions)
            ? currentStep.data.interruptions.map((item) => item instanceof RunToolApprovalItem
                ? serializeRunItem(item, agentIdentityKeys)
                : item)
            : currentStep.data?.interruptions;
        return {
            ...currentStep,
            data: {
                ...currentStep.data,
                interruptions,
            },
        };
    }
    return currentStep;
}
function serializeProcessedResponse(processedResponse, agentIdentityKeys) {
    return {
        ...processedResponse,
        newItems: processedResponse.newItems.map((item) => serializeRunItem(item, agentIdentityKeys)),
    };
}
/**
 * @internal
 */
export function deserializeSpan(trace, serializedSpan) {
    const spanData = serializedSpan.span_data;
    const previousSpan = serializedSpan.previous_span
        ? deserializeSpan(trace, serializedSpan.previous_span)
        : undefined;
    const span = getGlobalTraceProvider().createSpan({
        spanId: serializedSpan.id,
        traceId: serializedSpan.trace_id,
        parentId: serializedSpan.parent_id ?? undefined,
        startedAt: serializedSpan.started_at ?? undefined,
        endedAt: serializedSpan.ended_at ?? undefined,
        data: spanData,
    }, trace);
    span.previousSpan = previousSpan;
    return span;
}
/**
 * @internal
 */
export function deserializeModelResponse(serializedModelResponse) {
    const usage = new Usage(serializedModelResponse.usage);
    return {
        usage,
        output: serializedModelResponse.output.map((item) => protocol.OutputModelItem.parse(item)),
        responseId: serializedModelResponse.responseId,
        requestId: serializedModelResponse.requestId,
        providerData: serializedModelResponse.providerData,
    };
}
/**
 * @internal
 */
export function deserializeItem(serializedItem, agentMap) {
    switch (serializedItem.type) {
        case 'message_output_item':
            return new RunMessageOutputItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap));
        case 'tool_search_call_item':
            return new RunToolSearchCallItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap));
        case 'tool_search_output_item':
            return new RunToolSearchOutputItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap));
        case 'tool_call_item':
            return new RunToolCallItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap));
        case 'tool_call_output_item':
            return new RunToolCallOutputItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap), serializedItem.output, serializedItem.customData);
        case 'reasoning_item':
            return new RunReasoningItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap));
        case 'handoff_call_item':
            return new RunHandoffCallItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap));
        case 'handoff_output_item':
            return new RunHandoffOutputItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.sourceAgent, agentMap), resolveSerializedAgent(serializedItem.targetAgent, agentMap));
        case 'tool_approval_item':
            return new RunToolApprovalItem(serializedItem.rawItem, resolveSerializedAgent(serializedItem.agent, agentMap), serializedItem.toolName);
    }
}
function deserializeInterruptionItem(serializedItem, agentMap, currentAgent) {
    if (serializedItem instanceof RunToolApprovalItem) {
        return serializedItem;
    }
    const parsed = itemSchema.safeParse(serializedItem);
    if (parsed.success) {
        if (parsed.data.type === 'tool_approval_item') {
            const mappedAgent = resolveSerializedAgent(parsed.data.agent, agentMap, currentAgent);
            return new RunToolApprovalItem(parsed.data.rawItem, mappedAgent, parsed.data.toolName);
        }
        const item = deserializeItem(parsed.data, agentMap);
        return item instanceof RunToolApprovalItem ? item : undefined;
    }
    if (!serializedItem || typeof serializedItem !== 'object') {
        return undefined;
    }
    const value = serializedItem;
    if (!value.rawItem || typeof value.rawItem !== 'object') {
        return undefined;
    }
    const rawItem = value.rawItem;
    if (rawItem.type !== 'function_call' &&
        rawItem.type !== 'hosted_tool_call' &&
        rawItem.type !== 'computer_call' &&
        rawItem.type !== 'shell_call' &&
        rawItem.type !== 'apply_patch_call') {
        return undefined;
    }
    const agentName = value.agent && typeof value.agent.name === 'string'
        ? value.agent.name
        : undefined;
    const agentIdentity = value.agent && typeof value.agent.identity === 'string'
        ? value.agent.identity
        : undefined;
    const mappedAgent = agentName || agentIdentity
        ? resolveSerializedAgent({
            name: agentName ?? currentAgent.name,
            identity: agentIdentity,
        }, agentMap, currentAgent)
        : currentAgent;
    const toolName = typeof value.toolName === 'string'
        ? value.toolName
        : typeof rawItem.name === 'string'
            ? rawItem.name
            : undefined;
    return new RunToolApprovalItem(value.rawItem, mappedAgent, toolName);
}
function deserializeInterruptions(serializedInterruptions, agentMap, currentAgent) {
    if (!Array.isArray(serializedInterruptions)) {
        return [];
    }
    return serializedInterruptions
        .map((item) => deserializeInterruptionItem(item, agentMap, currentAgent))
        .filter((item) => item instanceof RunToolApprovalItem);
}
/**
 * @internal
 */
async function deserializeProcessedResponse(agentMap, state, serializedProcessedResponse, options = {}) {
    const currentAgent = state._currentAgent;
    const configuredTools = options.executionTools ?? (await currentAgent.getAllTools(state._context));
    const allTools = [
        ...configuredTools,
        ...state.getToolSearchRuntimeTools(currentAgent),
    ];
    const baseAgentTools = currentAgent.tools;
    const allowSerializedExecutionToolPlaceholder = options.allowSerializedExecutionToolPlaceholder ?? true;
    const tools = new Map(allTools
        .filter((tool) => tool.type === 'function')
        .map((tool) => [getFunctionToolQualifiedName(tool) ?? tool.name, tool]));
    const computerTools = new Map(allTools
        .filter((tool) => tool.type === 'computer')
        .map((tool) => [tool.name, tool]));
    const resolveComputerTool = (toolName) => {
        const exactMatch = computerTools.get(toolName);
        if (exactMatch) {
            return exactMatch;
        }
        if (toolName === 'computer') {
            return computerTools.get('computer_use_preview');
        }
        if (toolName === 'computer_use_preview') {
            return computerTools.get('computer');
        }
        return undefined;
    };
    const shellTools = new Map(allTools
        .filter((tool) => tool.type === 'shell')
        .map((tool) => [tool.name, tool]));
    const applyPatchTools = new Map(allTools
        .filter((tool) => tool.type === 'apply_patch')
        .map((tool) => [tool.name, tool]));
    const handoffs = new Map(currentAgent.handoffs.map((entry) => {
        if (entry instanceof Agent) {
            return [entry.name, handoff(entry)];
        }
        return [entry.toolName, entry];
    }));
    const result = {
        newItems: serializedProcessedResponse.newItems.map((item) => deserializeItem(item, agentMap)),
        toolsUsed: serializedProcessedResponse.toolsUsed,
        handoffs: serializedProcessedResponse.handoffs.map((handoff) => {
            if (!handoffs.has(handoff.handoff.toolName)) {
                throw new UserError(`Handoff ${handoff.handoff.toolName} not found`);
            }
            return {
                toolCall: handoff.toolCall,
                handoff: handoffs.get(handoff.handoff.toolName),
            };
        }),
        functions: await Promise.all(serializedProcessedResponse.functions.map(async (functionCall) => {
            const toolIdentity = resolveFunctionToolCallName(functionCall.toolCall, tools) ??
                functionCall.tool.name;
            const resolvedTool = tools.get(toolIdentity) ??
                getSerializedFunctionToolPlaceholder({
                    agent: currentAgent,
                    baseAgentTools,
                    serializedTool: functionCall.tool,
                    toolCall: functionCall.toolCall,
                    toolIdentity,
                    allowSerializedExecutionToolPlaceholder,
                });
            if (!resolvedTool) {
                throw new UserError(`Tool ${toolIdentity} not found`);
            }
            return {
                toolCall: functionCall.toolCall,
                tool: resolvedTool,
            };
        })),
        functionToolsNotFound: serializedProcessedResponse.functionToolsNotFound ?? [],
        computerActions: serializedProcessedResponse.computerActions.map((computerAction) => {
            const toolName = computerAction.computer.name;
            const computerTool = resolveComputerTool(toolName) ??
                getSerializedComputerToolPlaceholder({
                    agent: currentAgent,
                    baseAgentTools,
                    serializedTool: computerAction.computer,
                    toolName,
                    allowSerializedExecutionToolPlaceholder,
                });
            if (!computerTool) {
                throw new UserError(`Computer tool ${toolName} not found`);
            }
            return {
                toolCall: computerAction.toolCall,
                computer: computerTool,
            };
        }),
        shellActions: (serializedProcessedResponse.shellActions ?? []).map((shellAction) => {
            const toolName = shellAction.shell.name;
            const shellTool = shellTools.get(toolName) ??
                getSerializedShellToolPlaceholder({
                    agent: currentAgent,
                    baseAgentTools,
                    serializedTool: shellAction.shell,
                    toolName,
                    allowSerializedExecutionToolPlaceholder,
                });
            if (!shellTool) {
                throw new UserError(`Shell tool ${toolName} not found`);
            }
            return {
                toolCall: shellAction.toolCall,
                shell: shellTool,
            };
        }),
        applyPatchActions: (serializedProcessedResponse.applyPatchActions ?? []).map((applyPatchAction) => {
            const toolName = applyPatchAction.applyPatch.name;
            const applyPatchTool = applyPatchTools.get(toolName) ??
                getSerializedApplyPatchToolPlaceholder({
                    agent: currentAgent,
                    baseAgentTools,
                    serializedTool: applyPatchAction.applyPatch,
                    toolName,
                    allowSerializedExecutionToolPlaceholder,
                });
            if (!applyPatchTool) {
                throw new UserError(`Apply patch tool ${toolName} not found`);
            }
            return {
                toolCall: applyPatchAction.toolCall,
                applyPatch: applyPatchTool,
            };
        }),
        mcpApprovalRequests: (serializedProcessedResponse.mcpApprovalRequests ?? []).map((approvalRequest) => ({
            requestItem: new RunToolApprovalItem(approvalRequest.requestItem
                .rawItem, currentAgent),
            mcpTool: approvalRequest.mcpTool,
        })),
    };
    return {
        ...result,
        hasToolsOrApprovalsToRun() {
            return (result.handoffs.length > 0 ||
                result.functions.length > 0 ||
                result.functionToolsNotFound.length > 0 ||
                result.mcpApprovalRequests.length > 0 ||
                result.computerActions.length > 0 ||
                result.shellActions.length > 0 ||
                result.applyPatchActions.length > 0);
        },
    };
}
