// @openai/agents-core/dist/sandbox/runtime/livePreservedSessions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rememberLivePreservedOwnedSessions = rememberLivePreservedOwnedSessions;
exports.forgetLivePreservedOwnedSessions = forgetLivePreservedOwnedSessions;
exports.livePreservedOwnedSessionEntries = livePreservedOwnedSessionEntries;
exports.preservedOwnedSessionAgentKeysWithoutLiveReuse = preservedOwnedSessionAgentKeysWithoutLiveReuse;
exports.livePreservedOwnedSession = livePreservedOwnedSession;
const livePreservedOwnedSessionsByRunState = new WeakMap();
function rememberLivePreservedOwnedSessions(args) {
    const liveSessions = new Map();
    for (const [agentKey, entry] of Object.entries(args.serializedState.sessionsByAgent ?? {})) {
        if (!entry.preservedOwnedSession) {
            continue;
        }
        if (entry.reuseLiveSession === false) {
            continue;
        }
        const session = args.sessionsByAgentKey.get(agentKey);
        if (!session) {
            continue;
        }
        liveSessions.set(agentKey, {
            agentKey,
            backendId: entry.backendId,
            currentAgentName: entry.currentAgentName,
            session,
        });
    }
    if (liveSessions.size > 0) {
        livePreservedOwnedSessionsByRunState.set(args.state, liveSessions);
    }
    else {
        livePreservedOwnedSessionsByRunState.delete(args.state);
    }
}
function forgetLivePreservedOwnedSessions(state) {
    livePreservedOwnedSessionsByRunState.delete(state);
}
function livePreservedOwnedSessionEntries(state) {
    return [...(livePreservedOwnedSessionsByRunState.get(state)?.values() ?? [])];
}
function preservedOwnedSessionAgentKeysWithoutLiveReuse(serializedState) {
    return new Set(Object.entries(serializedState.sessionsByAgent ?? {})
        .filter(([, entry]) => entry.preservedOwnedSession && entry.reuseLiveSession === false)
        .map(([agentKey]) => agentKey));
}
function livePreservedOwnedSession(args) {
    if (!args.serializedEntry?.preservedOwnedSession || !args.runState) {
        return undefined;
    }
    const liveSessions = livePreservedOwnedSessionsByRunState.get(args.runState);
    const liveEntry = liveSessions?.get(args.agentKey);
    if (!liveEntry || liveEntry.backendId !== args.client.backendId) {
        return undefined;
    }
    return liveEntry;
}
