// @openai/agents-core/dist/sandbox/runtime/runAsManifest.mjs
import { Environment, Manifest } from "../manifest.mjs";
export function sandboxRunAsName(runAs) {
    if (runAs === undefined) {
        return undefined;
    }
    if (typeof runAs === 'string') {
        return runAs.trim();
    }
    return runAs.name;
}
export function manifestWithRunAsUser(manifest, runAs) {
    const runAsName = sandboxRunAsName(runAs);
    if (!runAsName || manifestHasUser(manifest, runAsName)) {
        return manifest;
    }
    return new Manifest({
        version: manifest.version,
        root: manifest.root,
        entries: structuredClone(manifest.entries),
        environment: Object.fromEntries(Object.entries(manifest.environment).map(([key, value]) => [
            key,
            value instanceof Environment ? value.init() : value,
        ])),
        users: [...structuredClone(manifest.users), { name: runAsName }],
        groups: structuredClone(manifest.groups),
        extraPathGrants: structuredClone(manifest.extraPathGrants),
        remoteMountCommandAllowlist: [...manifest.remoteMountCommandAllowlist],
    });
}
function manifestHasUser(manifest, name) {
    if (manifest.users.some((user) => user.name === name)) {
        return true;
    }
    return manifest.groups.some((group) => group.users?.some((user) => user.name === name));
}
