// @openai/agents-core/dist/sandbox/runtime/sessionState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSerializedSandboxState = getSerializedSandboxState;
exports.toSessionStateEnvelope = toSessionStateEnvelope;
exports.assertSessionStateEnvelope = assertSessionStateEnvelope;
exports.deserializeSandboxSessionStateEntry = deserializeSandboxSessionStateEntry;
exports.getSerializedSessionEntryForAgent = getSerializedSessionEntryForAgent;
exports.getPreviousSerializedSessionsByAgent = getPreviousSerializedSessionsByAgent;
exports.hasPreservedOwnedSessions = hasPreservedOwnedSessions;
const errors_1 = require("../../errors.js");
const session_1 = require("../session.js");
const manifestPersistence_1 = require("../sandboxes/shared/manifestPersistence.js");
function getSerializedSandboxState(runState) {
    return runState?._sandbox;
}
function toSessionStateEnvelope(backendId, state, providerState) {
    const persistentManifest = (0, manifestPersistence_1.serializeManifest)(state.manifest);
    return {
        version: session_1.SANDBOX_SESSION_STATE_VERSION,
        backendId,
        manifest: (0, manifestPersistence_1.serializeManifestRecord)(persistentManifest),
        ...(state.snapshot !== undefined ? { snapshot: state.snapshot } : {}),
        ...(state.snapshotFingerprint !== undefined
            ? { snapshotFingerprint: state.snapshotFingerprint }
            : {}),
        ...(state.snapshotFingerprintVersion !== undefined
            ? { snapshotFingerprintVersion: state.snapshotFingerprintVersion }
            : {}),
        workspaceReady: state.workspaceReady ?? true,
        ...(state.exposedPorts
            ? { exposedPorts: structuredClone(state.exposedPorts) }
            : {}),
        // Keep provider-owned fields separate from the SDK envelope so version and backend
        // checks can run before deserializing provider-specific state.
        providerState,
    };
}
function assertSessionStateEnvelope(client, envelope) {
    if (envelope.version !== session_1.SANDBOX_SESSION_STATE_VERSION) {
        throw new errors_1.UserError(`Sandbox session state version ${envelope.version} is not supported. Please use version ${session_1.SANDBOX_SESSION_STATE_VERSION}.`);
    }
    if (envelope.backendId !== client.backendId) {
        throw new errors_1.UserError('RunState sandbox session backend does not match the configured sandbox client.');
    }
}
async function deserializeSandboxSessionStateEntry(client, serializedEntry) {
    if (!serializedEntry) {
        return undefined;
    }
    if (serializedEntry.backendId !== client.backendId) {
        throw new errors_1.UserError('RunState sandbox backend does not match the configured sandbox client.');
    }
    if (!client.deserializeSessionState) {
        throw new errors_1.UserError('Sandbox client must implement deserializeSessionState() to resume RunState sandbox state.');
    }
    const envelope = serializedEntry.sessionState;
    assertSessionStateEnvelope(client, envelope);
    return await client.deserializeSessionState(providerStateWithSdkEnvelopeFields(envelope));
}
function providerStateWithSdkEnvelopeFields(envelope) {
    return {
        ...envelope.providerState,
        manifest: envelope.manifest,
        ...(envelope.snapshot !== undefined ? { snapshot: envelope.snapshot } : {}),
        ...(envelope.snapshotFingerprint !== undefined
            ? { snapshotFingerprint: envelope.snapshotFingerprint }
            : {}),
        ...(envelope.snapshotFingerprintVersion !== undefined
            ? {
                snapshotFingerprintVersion: envelope.snapshotFingerprintVersion,
            }
            : {}),
        workspaceReady: envelope.workspaceReady,
        ...(envelope.exposedPorts
            ? { exposedPorts: structuredClone(envelope.exposedPorts) }
            : {}),
    };
}
function getSerializedSessionEntryForAgent(sandboxState, currentAgentKey) {
    if (!sandboxState) {
        return undefined;
    }
    return (sandboxState.sessionsByAgent?.[currentAgentKey] ??
        (sandboxState.currentAgentKey === currentAgentKey
            ? {
                backendId: sandboxState.backendId,
                currentAgentKey: sandboxState.currentAgentKey,
                currentAgentName: sandboxState.currentAgentName,
                sessionState: sandboxState.sessionState,
            }
            : undefined));
}
function getPreviousSerializedSessionsByAgent(sandboxState, client) {
    if (!sandboxState || sandboxState.backendId !== client.backendId) {
        return {};
    }
    const entries = {
        ...(sandboxState.sessionsByAgent ?? {}),
    };
    if (sandboxState.currentAgentKey && !entries[sandboxState.currentAgentKey]) {
        entries[sandboxState.currentAgentKey] = {
            backendId: sandboxState.backendId,
            currentAgentKey: sandboxState.currentAgentKey,
            currentAgentName: sandboxState.currentAgentName,
            sessionState: sandboxState.sessionState,
        };
    }
    return Object.fromEntries(Object.entries(entries).filter(([, entry]) => entry.backendId === client.backendId));
}
function hasPreservedOwnedSessions(sandboxState) {
    return Object.values(sandboxState?.sessionsByAgent ?? {}).some((entry) => entry.preservedOwnedSession);
}
