// @openai/agents-core/dist/sandbox/runtime/sessionState.mjs
import { UserError } from "../../errors.mjs";
import { SANDBOX_SESSION_STATE_VERSION, } from "../session.mjs";
import { serializeManifest, serializeManifestRecord, } from "../sandboxes/shared/manifestPersistence.mjs";
export function getSerializedSandboxState(runState) {
    return runState?._sandbox;
}
export function toSessionStateEnvelope(backendId, state, providerState) {
    const persistentManifest = serializeManifest(state.manifest);
    return {
        version: SANDBOX_SESSION_STATE_VERSION,
        backendId,
        manifest: serializeManifestRecord(persistentManifest),
        ...(state.snapshot !== undefined ? { snapshot: state.snapshot } : {}),
        ...(state.snapshotFingerprint !== undefined
            ? { snapshotFingerprint: state.snapshotFingerprint }
            : {}),
        ...(state.snapshotFingerprintVersion !== undefined
            ? { snapshotFingerprintVersion: state.snapshotFingerprintVersion }
            : {}),
        workspaceReady: state.workspaceReady ?? true,
        ...(state.exposedPorts
            ? { exposedPorts: structuredClone(state.exposedPorts) }
            : {}),
        // Keep provider-owned fields separate from the SDK envelope so version and backend
        // checks can run before deserializing provider-specific state.
        providerState,
    };
}
export function assertSessionStateEnvelope(client, envelope) {
    if (envelope.version !== SANDBOX_SESSION_STATE_VERSION) {
        throw new UserError(`Sandbox session state version ${envelope.version} is not supported. Please use version ${SANDBOX_SESSION_STATE_VERSION}.`);
    }
    if (envelope.backendId !== client.backendId) {
        throw new UserError('RunState sandbox session backend does not match the configured sandbox client.');
    }
}
export async function deserializeSandboxSessionStateEntry(client, serializedEntry) {
    if (!serializedEntry) {
        return undefined;
    }
    if (serializedEntry.backendId !== client.backendId) {
        throw new UserError('RunState sandbox backend does not match the configured sandbox client.');
    }
    if (!client.deserializeSessionState) {
        throw new UserError('Sandbox client must implement deserializeSessionState() to resume RunState sandbox state.');
    }
    const envelope = serializedEntry.sessionState;
    assertSessionStateEnvelope(client, envelope);
    return await client.deserializeSessionState(providerStateWithSdkEnvelopeFields(envelope));
}
function providerStateWithSdkEnvelopeFields(envelope) {
    return {
        ...envelope.providerState,
        manifest: envelope.manifest,
        ...(envelope.snapshot !== undefined ? { snapshot: envelope.snapshot } : {}),
        ...(envelope.snapshotFingerprint !== undefined
            ? { snapshotFingerprint: envelope.snapshotFingerprint }
            : {}),
        ...(envelope.snapshotFingerprintVersion !== undefined
            ? {
                snapshotFingerprintVersion: envelope.snapshotFingerprintVersion,
            }
            : {}),
        workspaceReady: envelope.workspaceReady,
        ...(envelope.exposedPorts
            ? { exposedPorts: structuredClone(envelope.exposedPorts) }
            : {}),
    };
}
export function getSerializedSessionEntryForAgent(sandboxState, currentAgentKey) {
    if (!sandboxState) {
        return undefined;
    }
    return (sandboxState.sessionsByAgent?.[currentAgentKey] ??
        (sandboxState.currentAgentKey === currentAgentKey
            ? {
                backendId: sandboxState.backendId,
                currentAgentKey: sandboxState.currentAgentKey,
                currentAgentName: sandboxState.currentAgentName,
                sessionState: sandboxState.sessionState,
            }
            : undefined));
}
export function getPreviousSerializedSessionsByAgent(sandboxState, client) {
    if (!sandboxState || sandboxState.backendId !== client.backendId) {
        return {};
    }
    const entries = {
        ...(sandboxState.sessionsByAgent ?? {}),
    };
    if (sandboxState.currentAgentKey && !entries[sandboxState.currentAgentKey]) {
        entries[sandboxState.currentAgentKey] = {
            backendId: sandboxState.backendId,
            currentAgentKey: sandboxState.currentAgentKey,
            currentAgentName: sandboxState.currentAgentName,
            sessionState: sandboxState.sessionState,
        };
    }
    return Object.fromEntries(Object.entries(entries).filter(([, entry]) => entry.backendId === client.backendId));
}
export function hasPreservedOwnedSessions(sandboxState) {
    return Object.values(sandboxState?.sessionsByAgent ?? {}).some((entry) => entry.preservedOwnedSession);
}
