// @openai/agents-core/dist/sandbox/runtime/spans.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withSandboxSpan = withSandboxSpan;
const tracing_1 = require("../../tracing/index.js");
const events_1 = require("../events.js");
async function withSandboxSpan(name, data, fn) {
    const startedAt = Date.now();
    await (0, events_1.emitSandboxEvent)({
        type: 'sandbox_operation',
        name,
        phase: 'start',
        timestamp: new Date(startedAt).toISOString(),
        data: { ...data },
    });
    const runWithEvents = async (span) => {
        try {
            const result = await fn();
            await (0, events_1.emitSandboxEvent)({
                type: 'sandbox_operation',
                name,
                phase: 'end',
                timestamp: new Date().toISOString(),
                data: { ...data },
                durationMs: Date.now() - startedAt,
            });
            return result;
        }
        catch (error) {
            const serializedError = (0, events_1.serializeSandboxEventError)(error);
            recordSandboxSpanError(span, serializedError);
            await (0, events_1.emitSandboxEvent)({
                type: 'sandbox_operation',
                name,
                phase: 'error',
                timestamp: new Date().toISOString(),
                data: { ...data },
                durationMs: Date.now() - startedAt,
                error: serializedError,
            });
            throw error;
        }
    };
    if (!(0, tracing_1.getCurrentTrace)()) {
        return await runWithEvents();
    }
    return await (0, tracing_1.withCustomSpan)(async (span) => await runWithEvents(span), {
        data: {
            name,
            data,
        },
    });
}
function recordSandboxSpanError(span, error) {
    if (!span) {
        return;
    }
    span.spanData.data = {
        ...span.spanData.data,
        error,
        error_retryable: error.retryable ?? null,
    };
}
