// @openai/agents-core/dist/sandbox/sandboxes/shared/localSnapshots.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOCAL_SNAPSHOT_FINGERPRINT_VERSION = void 0;
exports.createLocalSnapshot = createLocalSnapshot;
exports.createRemoteSnapshot = createRemoteSnapshot;
exports.persistLocalSnapshot = persistLocalSnapshot;
exports.serializeLocalSnapshotSpec = serializeLocalSnapshotSpec;
exports.rehydrateLocalSnapshotSpec = rehydrateLocalSnapshotSpec;
exports.replaceLocalSnapshot = replaceLocalSnapshot;
exports.canReuseLocalSnapshotWorkspace = canReuseLocalSnapshotWorkspace;
exports.localSnapshotIsRestorable = localSnapshotIsRestorable;
exports.restoreLocalSnapshotToWorkspace = restoreLocalSnapshotToWorkspace;
exports.computeLocalSnapshotFingerprint = computeLocalSnapshotFingerprint;
exports.copyDirectory = copyDirectory;
exports.createWorkspaceArchive = createWorkspaceArchive;
exports.restoreWorkspaceArchive = restoreWorkspaceArchive;
exports.isRemoteSnapshotSpec = isRemoteSnapshotSpec;
const errors_1 = require("../../../errors.js");
const client_1 = require("../../client.js");
const errors_2 = require("../../errors.js");
const snapshot_1 = require("../../snapshot.js");
const promises_1 = require("node:fs/promises");
const node_fs_1 = require("node:fs");
const node_crypto_1 = require("node:crypto");
const node_path_1 = require("node:path");
const stableJson_1 = require("../../shared/stableJson.js");
const typeGuards_1 = require("../../shared/typeGuards.js");
const hostPath_1 = require("../../shared/hostPath.js");
const localSnapshotPaths_1 = require("./localSnapshotPaths.js");
const manifestPersistence_1 = require("./manifestPersistence.js");
exports.LOCAL_SNAPSHOT_FINGERPRINT_VERSION = 'workspace_tree_sha256_v1';
const SNAPSHOT_FILE_READ_FLAGS = node_fs_1.constants.O_RDONLY | node_fs_1.constants.O_NOFOLLOW;
const SNAPSHOT_FILE_WRITE_FLAGS = node_fs_1.constants.O_WRONLY |
    node_fs_1.constants.O_CREAT |
    node_fs_1.constants.O_TRUNC |
    node_fs_1.constants.O_NOFOLLOW;
async function createLocalSnapshot(state, spec) {
    const snapshotPath = await mkdtempForSnapshot(spec);
    await copyDirectory(state.workspaceRootPath, snapshotPath, localSnapshotExcludedPaths(state));
    return {
        id: (0, node_crypto_1.randomUUID)(),
        type: 'local',
        path: snapshotPath,
    };
}
async function createRemoteSnapshot(state, spec) {
    const data = await createWorkspaceArchive(state.workspaceRootPath, localSnapshotExcludedPaths(state));
    const saved = await spec.store.save({
        id: spec.id,
        data,
        metadata: spec.metadata,
    });
    return {
        id: saved.id,
        type: 'remote',
        ...(saved.metadata ? { metadata: saved.metadata } : {}),
    };
}
async function persistLocalSnapshot(providerName, state, spec) {
    if ((0, snapshot_1.isNoopSnapshotSpec)(spec)) {
        await replaceLocalSnapshot(state, null);
        replaceLocalSnapshotFingerprint(state, null);
        return null;
    }
    if (isRemoteSnapshotSpec(spec)) {
        const fingerprint = await computeLocalSnapshotFingerprint(state);
        const snapshot = await createRemoteSnapshot(state, spec);
        await replaceLocalSnapshot(state, snapshot);
        replaceLocalSnapshotFingerprint(state, fingerprint);
        return snapshot;
    }
    if (spec && spec.type !== 'local') {
        throw new errors_1.UserError(`${providerName} does not support snapshot specs of type "${spec.type}".`);
    }
    const fingerprint = await computeLocalSnapshotFingerprint(state);
    const snapshot = await createLocalSnapshot(state, spec ?? undefined);
    await replaceLocalSnapshot(state, snapshot);
    replaceLocalSnapshotFingerprint(state, fingerprint);
    return snapshot;
}
function serializeLocalSnapshotSpec(spec) {
    if (!spec) {
        return null;
    }
    if (spec.type === 'remote') {
        return {
            type: 'remote',
            ...(typeof spec.id === 'string' ? { id: spec.id } : {}),
            ...(spec.metadata ? { metadata: structuredClone(spec.metadata) } : {}),
        };
    }
    return structuredClone(spec);
}
function rehydrateLocalSnapshotSpec(serialized, configured) {
    // Remote SnapshotStore implementations can carry functions or credentials, so
    // serialized state stores only lookup fields and reattaches the live configured store.
    if (!(0, typeGuards_1.isRecord)(serialized)) {
        return null;
    }
    if (serialized.type !== 'remote') {
        return structuredClone(serialized);
    }
    if (isRemoteSnapshotSpec(serialized)) {
        return serialized;
    }
    if (!isRemoteSnapshotSpec(configured)) {
        return null;
    }
    const id = typeof serialized.id === 'string' ? serialized.id : configured.id;
    const metadata = (0, typeGuards_1.isRecord)(serialized.metadata)
        ? structuredClone(serialized.metadata)
        : configured.metadata;
    return {
        type: 'remote',
        store: configured.store,
        ...(id ? { id } : {}),
        ...(metadata ? { metadata } : {}),
    };
}
async function replaceLocalSnapshot(state, snapshot) {
    const previousSnapshot = state.snapshot;
    state.snapshot = snapshot;
    if (previousSnapshot?.type === 'local' &&
        previousSnapshot.path !== snapshot?.path &&
        previousSnapshot.path !== state.workspaceRootPath) {
        await (0, promises_1.rm)(previousSnapshot.path, { recursive: true, force: true });
    }
    if (previousSnapshot?.type === 'remote' &&
        previousSnapshot.id !== snapshot?.id) {
        const store = remoteSnapshotStoreFromSpec(state);
        await store?.delete?.({ id: previousSnapshot.id }).catch(() => undefined);
    }
}
async function canReuseLocalSnapshotWorkspace(state) {
    if (!(await localSnapshotIsRestorable(state))) {
        return true;
    }
    const storedFingerprint = state.snapshotFingerprint;
    const storedVersion = state.snapshotFingerprintVersion;
    if (!storedFingerprint ||
        storedVersion !== exports.LOCAL_SNAPSHOT_FINGERPRINT_VERSION) {
        return false;
    }
    const current = await computeLocalSnapshotFingerprint(state).catch(() => null);
    // The fingerprint covers the persisted manifest plus durable workspace tree; ephemeral
    // paths are intentionally ignored because they are rebuilt after resume.
    return (current?.fingerprint === storedFingerprint &&
        current.version === storedVersion);
}
async function localSnapshotIsRestorable(state) {
    if (state.snapshot?.type === 'remote') {
        const store = remoteSnapshotStoreFromSpec(state);
        if (!store) {
            return false;
        }
        if (store.exists) {
            return await store.exists({ id: state.snapshot.id });
        }
        return true;
    }
    return Boolean(state.snapshot?.type === 'local' && (await pathExists(state.snapshot.path)));
}
async function restoreLocalSnapshotToWorkspace(state, workspaceRootPath, options = {}) {
    if (!(await localSnapshotIsRestorable(state))) {
        throw new errors_1.UserError('No local snapshot is available to restore.');
    }
    if (state.snapshot?.type === 'remote') {
        const store = remoteSnapshotStoreFromSpec(state);
        if (!store) {
            throw new errors_1.UserError('No remote snapshot store is available to restore.');
        }
        const restored = await store.load({ id: state.snapshot.id });
        await restoreWorkspaceArchive(restored.data, workspaceRootPath, options);
    }
    else {
        await replaceDirectoryContents(state.snapshot.path, workspaceRootPath);
    }
    state.workspaceRootPath = workspaceRootPath;
    const fingerprint = await computeLocalSnapshotFingerprint(state).catch(() => null);
    replaceLocalSnapshotFingerprint(state, fingerprint);
    return state;
}
async function computeLocalSnapshotFingerprint(state) {
    const hash = (0, node_crypto_1.createHash)('sha256');
    appendHashFrame(hash, 'version', exports.LOCAL_SNAPSHOT_FINGERPRINT_VERSION);
    appendHashFrame(hash, 'manifest', (0, stableJson_1.stableJsonStringify)((0, manifestPersistence_1.serializeManifest)(state.manifest), {
        encodeBytes: (value) => ({
            type: 'Uint8Array',
            data: Buffer.from(value).toString('base64'),
        }),
    }));
    await hashDirectory(hash, state.workspaceRootPath, localSnapshotExcludedPaths(state));
    return {
        fingerprint: hash.digest('hex'),
        version: exports.LOCAL_SNAPSHOT_FINGERPRINT_VERSION,
    };
}
function localSnapshotExcludedPaths(state) {
    const paths = state.manifest.ephemeralPersistencePaths();
    for (const path of state.snapshotExcludedPaths ?? []) {
        paths.add(path);
    }
    return paths;
}
async function copyDirectory(sourceRoot, destinationRoot, excludedLogicalPaths = new Set(), currentRelativePath = '', expectedSourceStats, destinationRootPath = destinationRoot) {
    const entries = await readStableSnapshotDirectoryEntries(sourceRoot, expectedSourceStats);
    await assertSafeSnapshotDestinationPath(destinationRootPath, destinationRoot);
    await (0, promises_1.mkdir)(destinationRoot, { recursive: true });
    await assertSafeSnapshotDestinationPath(destinationRootPath, destinationRoot);
    for (const entry of entries) {
        const relativePath = currentRelativePath
            ? `${currentRelativePath}/${entry.name}`
            : entry.name;
        if (shouldSkipSnapshotPath(relativePath, excludedLogicalPaths)) {
            continue;
        }
        const sourcePath = (0, node_path_1.join)(sourceRoot, entry.name);
        const destinationPath = (0, node_path_1.join)(destinationRoot, entry.name);
        const sourceStats = await (0, promises_1.lstat)(sourcePath);
        if (sourceStats.isDirectory()) {
            await copyDirectory(sourcePath, destinationPath, excludedLogicalPaths, relativePath, sourceStats, destinationRootPath);
            continue;
        }
        if (!sourceStats.isFile()) {
            continue;
        }
        await assertSafeSnapshotDestinationPath(destinationRootPath, (0, node_path_1.dirname)(destinationPath));
        await (0, promises_1.mkdir)((0, node_path_1.dirname)(destinationPath), { recursive: true });
        await assertSafeSnapshotDestinationPath(destinationRootPath, (0, node_path_1.dirname)(destinationPath));
        await writeStableSnapshotFile(destinationRootPath, destinationPath, await readStableSnapshotFile(sourcePath, sourceStats), sourceStats.mode & 0o777);
        const destinationStats = await (0, promises_1.lstat)(destinationPath);
        if (!destinationStats.isFile()) {
            await (0, promises_1.rm)(destinationPath, { recursive: true, force: true });
        }
    }
}
async function createWorkspaceArchive(sourceRoot, excludedLogicalPaths) {
    // Remote SnapshotStore is provider-agnostic, so its portable fallback is a small
    // JSON/base64 archive instead of relying on host tar availability.
    const archive = {
        version: 1,
        directories: [],
        files: [],
    };
    await appendWorkspaceArchiveEntries(archive, sourceRoot, excludedLogicalPaths);
    return new TextEncoder().encode(JSON.stringify(archive));
}
async function appendWorkspaceArchiveEntries(archive, sourceRoot, excludedLogicalPaths, currentRelativePath = '', expectedSourceStats) {
    const entries = await readStableSnapshotDirectoryEntries(sourceRoot, expectedSourceStats);
    for (const entry of entries.sort((left, right) => left.name.localeCompare(right.name))) {
        const relativePath = currentRelativePath
            ? `${currentRelativePath}/${entry.name}`
            : entry.name;
        if (shouldSkipSnapshotPath(relativePath, excludedLogicalPaths)) {
            continue;
        }
        const sourcePath = (0, node_path_1.join)(sourceRoot, entry.name);
        const sourceStats = await (0, promises_1.lstat)(sourcePath);
        if (sourceStats.isDirectory()) {
            archive.directories.push(relativePath);
            await appendWorkspaceArchiveEntries(archive, sourcePath, excludedLogicalPaths, relativePath, sourceStats);
            continue;
        }
        if (!sourceStats.isFile()) {
            continue;
        }
        archive.files.push({
            path: relativePath,
            data: Buffer.from(await readStableSnapshotFile(sourcePath, sourceStats)).toString('base64'),
        });
    }
}
async function restoreWorkspaceArchive(data, destinationRoot, options = {}) {
    const bytes = typeof data === 'string'
        ? new TextEncoder().encode(data)
        : workspaceArchiveDataToUint8Array(data);
    const archiveLimits = (0, client_1.resolveSandboxArchiveLimits)(options.archiveLimits);
    checkWorkspaceArchiveInputBytes(bytes.byteLength, archiveLimits);
    const archive = JSON.parse(new TextDecoder().decode(bytes));
    if (archive.version !== 1) {
        throw new errors_1.UserError('Unsupported remote snapshot archive version.');
    }
    validateWorkspaceArchiveLimits(archive, archiveLimits);
    await (0, promises_1.mkdir)(destinationRoot, { recursive: true });
    await clearDirectory(destinationRoot);
    for (const directory of archive.directories) {
        await writeSafeArchiveDirectory(destinationRoot, directory);
    }
    for (const file of archive.files) {
        await writeSafeArchiveFile(destinationRoot, file.path, file.data);
    }
}
function workspaceArchiveDataToUint8Array(data) {
    if (data instanceof Uint8Array) {
        return data;
    }
    return new Uint8Array(data);
}
function checkWorkspaceArchiveInputBytes(actual, limits) {
    const limit = limits?.maxInputBytes;
    if (limit != null && actual > limit) {
        throw new errors_2.SandboxArchiveError('Workspace archive input size exceeds limit.', {
            reason: 'archive input size exceeds limit',
            limit,
            actual,
        });
    }
}
function validateWorkspaceArchiveLimits(archive, limits) {
    if (limits == null) {
        return;
    }
    let memberCount = 0;
    for (const directory of archive.directories) {
        memberCount += 1;
        checkWorkspaceArchiveMemberCount(memberCount, directory, limits);
    }
    let extractedBytes = 0;
    for (const file of archive.files) {
        memberCount += 1;
        checkWorkspaceArchiveMemberCount(memberCount, file.path, limits);
        extractedBytes += Buffer.byteLength(file.data, 'base64');
        checkWorkspaceArchiveExtractedBytes(extractedBytes, file.path, limits);
    }
}
function checkWorkspaceArchiveMemberCount(actual, member, limits) {
    const limit = limits?.maxMembers;
    if (limit != null && actual > limit) {
        throw new errors_2.SandboxArchiveError('Workspace archive member count exceeds limit.', {
            reason: 'archive member count exceeds limit',
            limit,
            actual,
            member,
        });
    }
}
function checkWorkspaceArchiveExtractedBytes(actual, member, limits) {
    const limit = limits?.maxExtractedBytes;
    if (limit != null && actual > limit) {
        throw new errors_2.SandboxArchiveError('Workspace archive extracted size exceeds limit.', {
            reason: 'archive extracted size exceeds limit',
            limit,
            actual,
            member,
        });
    }
}
async function writeSafeArchiveDirectory(destinationRoot, relativePath) {
    const destinationPath = safeArchivePath(destinationRoot, relativePath);
    await assertSafeSnapshotDestinationPath(destinationRoot, destinationPath);
    await (0, promises_1.mkdir)(destinationPath, { recursive: true });
    await assertSafeSnapshotDestinationPath(destinationRoot, destinationPath);
}
async function writeSafeArchiveFile(destinationRoot, relativePath, base64Data) {
    const destinationPath = safeArchivePath(destinationRoot, relativePath);
    await assertSafeSnapshotDestinationPath(destinationRoot, (0, node_path_1.dirname)(destinationPath));
    await (0, promises_1.mkdir)((0, node_path_1.dirname)(destinationPath), { recursive: true });
    await assertSafeSnapshotDestinationPath(destinationRoot, (0, node_path_1.dirname)(destinationPath));
    await writeStableSnapshotFile(destinationRoot, destinationPath, Buffer.from(base64Data, 'base64'));
}
function safeArchivePath(destinationRoot, relativePath) {
    const destinationPath = (0, node_path_1.join)(destinationRoot, relativePath);
    if (!(0, hostPath_1.isHostPathStrictlyWithinRoot)(destinationRoot, destinationPath)) {
        throw new errors_1.UserError(`Remote snapshot archive path escapes root: ${relativePath}`);
    }
    return destinationPath;
}
async function replaceDirectoryContents(sourceRoot, destinationRoot) {
    await assertSafeSnapshotDestinationPath(destinationRoot, destinationRoot);
    await (0, promises_1.mkdir)(destinationRoot, { recursive: true });
    await assertSafeSnapshotDestinationPath(destinationRoot, destinationRoot);
    await clearDirectory(destinationRoot);
    await copyDirectory(sourceRoot, destinationRoot);
}
async function clearDirectory(destinationRoot) {
    await assertSafeSnapshotDestinationPath(destinationRoot, destinationRoot);
    const entries = await (0, promises_1.readdir)(destinationRoot);
    await Promise.all(entries.map(async (entry) => {
        await (0, promises_1.rm)((0, node_path_1.join)(destinationRoot, entry), { recursive: true, force: true });
    }));
}
function replaceLocalSnapshotFingerprint(state, fingerprint) {
    state.snapshotFingerprint = fingerprint?.fingerprint ?? null;
    state.snapshotFingerprintVersion = fingerprint?.version ?? null;
}
async function mkdtempForSnapshot(spec) {
    const baseDir = spec?.baseDir ?? (0, localSnapshotPaths_1.defaultLocalSnapshotBaseDir)();
    await (0, promises_1.mkdir)(baseDir, { recursive: true });
    return await (0, promises_1.mkdtemp)((0, node_path_1.join)(baseDir, 'openai-agents-sandbox-snapshot-'));
}
function shouldSkipSnapshotPath(relativePath, excludedLogicalPaths) {
    for (const excluded of excludedLogicalPaths) {
        if (excluded === '' ||
            relativePath === excluded ||
            relativePath.startsWith(`${excluded}/`)) {
            return true;
        }
    }
    return false;
}
function isRemoteSnapshotSpec(spec) {
    const store = spec?.store;
    return (spec?.type === 'remote' &&
        typeof store === 'object' &&
        store !== null &&
        typeof store.save === 'function' &&
        typeof store.load === 'function');
}
function remoteSnapshotStoreFromSpec(state) {
    const spec = state.snapshotSpec;
    return isRemoteSnapshotSpec(spec) ? spec.store : undefined;
}
async function pathExists(path) {
    try {
        await (0, promises_1.stat)(path);
        return true;
    }
    catch {
        return false;
    }
}
async function hashDirectory(hash, rootPath, excludedLogicalPaths, currentRelativePath = '', expectedSourceStats) {
    const entries = await readStableSnapshotDirectoryEntries(rootPath, expectedSourceStats);
    for (const entry of entries.sort((left, right) => left.name.localeCompare(right.name))) {
        const relativePath = currentRelativePath
            ? `${currentRelativePath}/${entry.name}`
            : entry.name;
        if (shouldSkipSnapshotPath(relativePath, excludedLogicalPaths)) {
            continue;
        }
        const entryPath = (0, node_path_1.join)(rootPath, entry.name);
        const entryStats = await (0, promises_1.lstat)(entryPath);
        if (entryStats.isDirectory()) {
            appendHashFrame(hash, 'dir', relativePath);
            await hashDirectory(hash, entryPath, excludedLogicalPaths, relativePath, entryStats);
            continue;
        }
        if (!entryStats.isFile()) {
            continue;
        }
        const content = await readStableSnapshotFile(entryPath, entryStats);
        appendHashFrame(hash, 'file', relativePath);
        appendHashFrame(hash, 'bytes', String(content.byteLength));
        hash.update(content);
    }
}
async function readStableSnapshotDirectoryEntries(sourceRoot, expectedSourceStats) {
    const sourceStats = await (0, promises_1.lstat)(sourceRoot);
    if (!sourceStats.isDirectory() ||
        (expectedSourceStats &&
            !sameFilesystemEntry(sourceStats, expectedSourceStats))) {
        throw snapshotPathChangedError(sourceRoot);
    }
    const entries = await (0, promises_1.readdir)(sourceRoot, { withFileTypes: true });
    await assertStableSnapshotDirectory(sourceRoot, sourceStats);
    return entries;
}
async function assertStableSnapshotDirectory(sourceRoot, expectedSourceStats) {
    const sourceStats = await (0, promises_1.lstat)(sourceRoot);
    if (!sourceStats.isDirectory() ||
        !sameFilesystemEntry(sourceStats, expectedSourceStats)) {
        throw snapshotPathChangedError(sourceRoot);
    }
}
async function readStableSnapshotFile(sourcePath, expectedSourceStats) {
    const sourceStats = await (0, promises_1.lstat)(sourcePath);
    if (!sourceStats.isFile() ||
        !sameFilesystemEntry(sourceStats, expectedSourceStats)) {
        throw snapshotPathChangedError(sourcePath);
    }
    let handle;
    try {
        handle = await (0, promises_1.open)(sourcePath, SNAPSHOT_FILE_READ_FLAGS);
    }
    catch (error) {
        if (isPathChangedError(error)) {
            throw snapshotPathChangedError(sourcePath);
        }
        throw error;
    }
    try {
        const openedStats = await handle.stat();
        if (!openedStats.isFile() ||
            !sameFilesystemEntry(openedStats, sourceStats)) {
            throw snapshotPathChangedError(sourcePath);
        }
        return await handle.readFile();
    }
    finally {
        await handle.close();
    }
}
async function writeStableSnapshotFile(destinationRoot, destinationPath, content, mode) {
    await assertSafeSnapshotDestinationPath(destinationRoot, destinationPath);
    let handle;
    try {
        handle = await (0, promises_1.open)(destinationPath, SNAPSHOT_FILE_WRITE_FLAGS, mode);
    }
    catch (error) {
        if (isPathChangedError(error)) {
            throw snapshotPathChangedError(destinationPath);
        }
        throw error;
    }
    try {
        await handle.writeFile(content);
        if (mode !== undefined) {
            await handle.chmod(mode);
        }
    }
    finally {
        await handle.close();
    }
    await assertSafeSnapshotDestinationPath(destinationRoot, destinationPath);
}
async function assertSafeSnapshotDestinationPath(destinationRoot, destinationPath) {
    const rootStats = await (0, promises_1.lstat)(destinationRoot).catch(() => undefined);
    if (!rootStats?.isDirectory() || rootStats.isSymbolicLink()) {
        throw snapshotPathChangedError(destinationRoot);
    }
    const rootRealPath = await (0, promises_1.realpath)(destinationRoot).catch(() => {
        throw snapshotPathChangedError(destinationRoot);
    });
    const relativeDestination = (0, node_path_1.relative)(destinationRoot, destinationPath);
    if ((0, hostPath_1.relativeHostPathEscapesRoot)(relativeDestination)) {
        throw snapshotPathChangedError(destinationPath);
    }
    if (relativeDestination === '') {
        return;
    }
    let current = destinationRoot;
    for (const segment of relativeDestination.split(node_path_1.sep)) {
        if (!segment) {
            continue;
        }
        current = (0, node_path_1.join)(current, segment);
        const currentStats = await (0, promises_1.lstat)(current).catch((error) => {
            if (error.code === 'ENOENT') {
                return undefined;
            }
            throw error;
        });
        if (!currentStats) {
            return;
        }
        if (currentStats.isSymbolicLink()) {
            throw snapshotPathChangedError(current);
        }
        const currentRealPath = await (0, promises_1.realpath)(current).catch(() => {
            throw snapshotPathChangedError(current);
        });
        if (!(0, hostPath_1.isHostPathWithinRoot)(rootRealPath, currentRealPath)) {
            throw snapshotPathChangedError(current);
        }
    }
}
function snapshotPathChangedError(path) {
    return new errors_1.UserError(`Sandbox snapshot path changed while copying: ${path}`);
}
function isPathChangedError(error) {
    return (typeof error === 'object' &&
        error !== null &&
        'code' in error &&
        (error.code === 'ELOOP' ||
            error.code === 'ENOENT' ||
            error.code === 'ENOTDIR'));
}
function sameFilesystemEntry(left, right) {
    return left.dev === right.dev && left.ino === right.ino;
}
function appendHashFrame(hash, label, value) {
    // Length-delimited frames avoid collisions such as ["ab", "c"] vs ["a", "bc"].
    hash.update(label);
    hash.update('\0');
    hash.update(String(value.length));
    hash.update('\0');
    hash.update(value);
    hash.update('\0');
}
