// @openai/agents-core/dist/sandbox/sandboxes/shared/manifestPersistence.mjs
import { isMount } from "../../entries/index.mjs";
import { Manifest } from "../../manifest.mjs";
import { mergeNamedObjects, mergePathKeyedObjects, } from "../../shared/manifestCollections.mjs";
import { hasCustomRemoteMountCommandAllowlist } from "../../shared/remoteMountCommandAllowlist.mjs";
import { serializeManifestEnvironment, } from "../../shared/environment.mjs";
import { encodeUint8ArrayToBase64 } from "../../../utils/base64.mjs";
export function serializeManifest(manifest) {
    return deserializeManifest(serializeManifestRecord(manifest));
}
export function serializeManifestRecord(manifest) {
    return {
        version: manifest.version,
        root: manifest.root,
        entries: sanitizeEntriesForPersistence(manifest.entries),
        environment: serializePersistentManifestEnvironment(manifest),
        users: structuredClone(manifest.users),
        groups: structuredClone(manifest.groups),
        extraPathGrants: structuredClone(manifest.extraPathGrants),
        remoteMountCommandAllowlist: [...manifest.remoteMountCommandAllowlist],
    };
}
export function deserializeManifest(value) {
    return new Manifest(deserializeManifestRecord(value ?? {}));
}
export function sanitizeEnvironmentForPersistence(state) {
    const environment = serializeEnvironmentForPersistence(state);
    return Object.fromEntries(Object.entries(environment).map(([key, value]) => [key, value.value]));
}
export function serializeEnvironmentForPersistence(state) {
    const runtimeEnvironment = state.environment ?? {};
    const ephemeralKeys = new Set();
    const serialized = {};
    for (const [key, value] of Object.entries(state.manifest.environment)) {
        if (value.ephemeral) {
            ephemeralKeys.add(key);
            continue;
        }
        serialized[key] = {
            ...value.normalized(),
            value: runtimeEnvironment[key] ?? value.value,
        };
    }
    for (const [key, value] of Object.entries(runtimeEnvironment)) {
        if (key in state.manifest.environment || ephemeralKeys.has(key)) {
            continue;
        }
        // Provider startup may add runtime env vars that are not in the manifest; keep them
        // unless they collide with an explicitly ephemeral manifest key.
        serialized[key] = { value };
    }
    return serialized;
}
export function mergeManifestDelta(base, update) {
    return new Manifest({
        version: update.version ?? base.version,
        root: base.root,
        entries: {
            ...structuredClone(base.entries),
            ...structuredClone(update.entries),
        },
        environment: {
            ...cloneManifestEnvironment(base),
            ...cloneManifestEnvironment(update),
        },
        users: mergeNamedObjects(base.users, update.users),
        groups: mergeNamedObjects(base.groups, update.groups),
        extraPathGrants: mergePathKeyedObjects(base.extraPathGrants, update.extraPathGrants),
        remoteMountCommandAllowlist: shouldMergeRemoteMountCommandAllowlist(update)
            ? update.remoteMountCommandAllowlist
            : base.remoteMountCommandAllowlist,
    });
}
function cloneManifestEnvironment(manifest) {
    return Object.fromEntries(Object.entries(manifest.environment).map(([key, value]) => [
        key,
        value.init(),
    ]));
}
export function mergeManifestEntryDelta(base, path, entry) {
    return mergeManifestDelta(base, new Manifest({
        root: base.root,
        entries: {
            [path]: structuredClone(entry),
        },
    }));
}
function serializePersistentManifestEnvironment(manifest) {
    const environment = serializeManifestEnvironment(manifest);
    // Ephemeral envs are runtime-only and must not be persisted.
    // Persisting them would leak values into snapshots.
    // Resume would also depend on regenerated data.
    return Object.fromEntries(Object.entries(environment).filter(([key]) => !manifest.environment[key]?.ephemeral));
}
function shouldMergeRemoteMountCommandAllowlist(update) {
    return hasCustomRemoteMountCommandAllowlist(update.remoteMountCommandAllowlist);
}
function sanitizeEntriesForPersistence(entries, ancestorEphemeral = false) {
    const sanitizedEntries = {};
    for (const [path, entry] of Object.entries(entries)) {
        const sanitizedEntry = sanitizeEntryForPersistence(entry, ancestorEphemeral);
        if (sanitizedEntry) {
            sanitizedEntries[path] = sanitizedEntry;
        }
    }
    return sanitizedEntries;
}
function sanitizeEntryForPersistence(entry, ancestorEphemeral) {
    const effectiveEphemeral = ancestorEphemeral || Boolean(entry.ephemeral);
    if (entry.type === 'file') {
        return !effectiveEphemeral
            ? {
                ...entry,
                content: serializeFileContentForPersistence(entry.content),
            }
            : undefined;
    }
    if (entry.type !== 'dir' || !entry.children) {
        return !effectiveEphemeral || isMount(entry) ? { ...entry } : undefined;
    }
    const children = sanitizeEntriesForPersistence(entry.children, effectiveEphemeral);
    if (effectiveEphemeral && Object.keys(children).length === 0) {
        return undefined;
    }
    return {
        ...entry,
        children,
    };
}
function deserializeManifestRecord(value) {
    return {
        ...value,
        entries: deserializeEntriesForRuntime(value.entries),
    };
}
function deserializeEntriesForRuntime(entries) {
    if (!entries) {
        return {};
    }
    return Object.fromEntries(Object.entries(entries).map(([path, entry]) => [
        path,
        deserializeEntryForRuntime(entry),
    ]));
}
function deserializeEntryForRuntime(entry) {
    if (entry.type === 'file') {
        const content = entry.content;
        return {
            ...entry,
            content: isSerializedFileContent(content)
                ? decodeBase64ToUint8Array(content.data)
                : entry.content,
        };
    }
    if (entry.type === 'dir' && entry.children) {
        return {
            ...entry,
            children: deserializeEntriesForRuntime(entry.children),
        };
    }
    return { ...entry };
}
function serializeFileContentForPersistence(content) {
    return typeof content === 'string'
        ? content
        : {
            type: 'base64',
            data: encodeUint8ArrayToBase64(content),
        };
}
function isSerializedFileContent(value) {
    return (typeof value === 'object' &&
        value !== null &&
        value.type === 'base64' &&
        typeof value.data === 'string');
}
function decodeBase64ToUint8Array(value) {
    const bufferCtor = globalThis.Buffer;
    if (bufferCtor) {
        return Uint8Array.from(bufferCtor.from(value, 'base64'));
    }
    const binary = atob(value);
    const bytes = new Uint8Array(binary.length);
    for (let index = 0; index < binary.length; index += 1) {
        bytes[index] = binary.charCodeAt(index);
    }
    return bytes;
}
