// @openai/agents-core/dist/sandbox/sandboxes/shared/sessionStateValues.mjs
import { Manifest } from "../../manifest.mjs";
import { normalizeExposedPort } from "../../session.mjs";
import { isRecord, isStringRecord, readOptionalNumberArray, readOptionalString, readString, } from "../../shared/typeGuards.mjs";
import { rehydrateLocalSnapshotSpec } from "./localSnapshots.mjs";
export function deserializeLocalSandboxSessionStateValues(state, configuredSnapshot) {
    return {
        manifest: new Manifest(state.manifest),
        workspaceRootPath: readString(state, 'workspaceRootPath'),
        workspaceRootOwned: Boolean(state.workspaceRootOwned),
        environment: readEnvironmentState(state.environment),
        snapshotSpec: rehydrateLocalSnapshotSpec(state.snapshotSpec, configuredSnapshot),
        snapshot: state.snapshot ?? null,
        snapshotFingerprint: readOptionalString(state, 'snapshotFingerprint') ??
            readOptionalString(state, 'snapshot_fingerprint') ??
            null,
        snapshotFingerprintVersion: readOptionalString(state, 'snapshotFingerprintVersion') ??
            readOptionalString(state, 'snapshot_fingerprint_version') ??
            null,
        configuredExposedPorts: normalizeExposedPorts(readOptionalNumberArray(state.configuredExposedPorts)),
        exposedPorts: readExposedPortsState(state),
    };
}
export function readExposedPortsState(state) {
    const exposedPorts = state.exposedPorts;
    return isRecord(exposedPorts)
        ? exposedPorts
        : undefined;
}
export function normalizeExposedPorts(ports) {
    return [...new Set((ports ?? []).map((port) => normalizeExposedPort(port)))];
}
function readEnvironmentState(value) {
    return isStringRecord(value) ? { ...value } : {};
}
