// @openai/agents-core/dist/sandbox/sandboxes/shared/sessionStateValues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeLocalSandboxSessionStateValues = deserializeLocalSandboxSessionStateValues;
exports.readExposedPortsState = readExposedPortsState;
exports.normalizeExposedPorts = normalizeExposedPorts;
const manifest_1 = require("../../manifest.js");
const session_1 = require("../../session.js");
const typeGuards_1 = require("../../shared/typeGuards.js");
const localSnapshots_1 = require("./localSnapshots.js");
function deserializeLocalSandboxSessionStateValues(state, configuredSnapshot) {
    return {
        manifest: new manifest_1.Manifest(state.manifest),
        workspaceRootPath: (0, typeGuards_1.readString)(state, 'workspaceRootPath'),
        workspaceRootOwned: Boolean(state.workspaceRootOwned),
        environment: readEnvironmentState(state.environment),
        snapshotSpec: (0, localSnapshots_1.rehydrateLocalSnapshotSpec)(state.snapshotSpec, configuredSnapshot),
        snapshot: state.snapshot ?? null,
        snapshotFingerprint: (0, typeGuards_1.readOptionalString)(state, 'snapshotFingerprint') ??
            (0, typeGuards_1.readOptionalString)(state, 'snapshot_fingerprint') ??
            null,
        snapshotFingerprintVersion: (0, typeGuards_1.readOptionalString)(state, 'snapshotFingerprintVersion') ??
            (0, typeGuards_1.readOptionalString)(state, 'snapshot_fingerprint_version') ??
            null,
        configuredExposedPorts: normalizeExposedPorts((0, typeGuards_1.readOptionalNumberArray)(state.configuredExposedPorts)),
        exposedPorts: readExposedPortsState(state),
    };
}
function readExposedPortsState(state) {
    const exposedPorts = state.exposedPorts;
    return (0, typeGuards_1.isRecord)(exposedPorts)
        ? exposedPorts
        : undefined;
}
function normalizeExposedPorts(ports) {
    return [...new Set((ports ?? []).map((port) => (0, session_1.normalizeExposedPort)(port)))];
}
function readEnvironmentState(value) {
    return (0, typeGuards_1.isStringRecord)(value) ? { ...value } : {};
}
