// @openai/agents-core/dist/sandbox/capabilities/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Capability = void 0;
exports.requireBoundSession = requireBoundSession;
const errors_1 = require("../errors.js");
class Capability {
    _session;
    _runAs;
    _modelInstance;
    clone() {
        const cloned = Object.create(Object.getPrototypeOf(this));
        for (const [key, value] of Object.entries(this)) {
            if (key === '_session' || key === '_runAs' || key === '_modelInstance') {
                continue;
            }
            cloned[key] = cloneCapabilityValue(value);
        }
        return cloned;
    }
    bind(session) {
        this._session = session;
        return this;
    }
    bindRunAs(runAs) {
        this._runAs = typeof runAs === 'string' ? runAs : runAs?.name;
        return this;
    }
    bindModel(_model, modelInstance) {
        this._modelInstance = modelInstance;
        return this;
    }
    requiredCapabilityTypes() {
        return new Set();
    }
    tools() {
        return [];
    }
    processManifest(manifest) {
        return manifest;
    }
    instructions(_manifest) {
        return null;
    }
    samplingParams(_samplingParams) {
        return {};
    }
    processContext(context) {
        return context;
    }
}
exports.Capability = Capability;
function requireBoundSession(capabilityType, session) {
    if (!session) {
        throw new errors_1.SandboxConfigurationError(`${capitalize(capabilityType)} capability is not bound to a SandboxSession`, { capability: capabilityType });
    }
    return session;
}
function cloneCapabilityValue(value) {
    if (Array.isArray(value)) {
        return value.map((item) => cloneCapabilityValue(item));
    }
    if (value instanceof Uint8Array) {
        return new Uint8Array(value);
    }
    if (value instanceof Map) {
        return new Map(value);
    }
    if (value instanceof Set) {
        return new Set(value);
    }
    if (isPlainRecord(value)) {
        return Object.fromEntries(Object.entries(value).map(([key, childValue]) => [
            key,
            cloneCapabilityValue(childValue),
        ]));
    }
    return value;
}
function isPlainRecord(value) {
    if (typeof value !== 'object' || value === null) {
        return false;
    }
    const prototype = Object.getPrototypeOf(value);
    return prototype === Object.prototype || prototype === null;
}
function capitalize(value) {
    return value.charAt(0).toUpperCase() + value.slice(1);
}
