// @openai/agents-core/dist/sandbox/capabilities/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Capabilities = exports.skills = exports.shell = exports.InMemoryMemoryStore = exports.memory = exports.filesystem = exports.compaction = exports.StaticCompactionPolicy = exports.DynamicCompactionPolicy = exports.CompactionPolicy = exports.CompactionModelInfo = exports.Capability = void 0;
const compaction_1 = require("./compaction.js");
const filesystem_1 = require("./filesystem.js");
const shell_1 = require("./shell.js");
var base_1 = require("./base.js");
Object.defineProperty(exports, "Capability", { enumerable: true, get: function () { return base_1.Capability; } });
var compaction_2 = require("./compaction.js");
Object.defineProperty(exports, "CompactionModelInfo", { enumerable: true, get: function () { return compaction_2.CompactionModelInfo; } });
Object.defineProperty(exports, "CompactionPolicy", { enumerable: true, get: function () { return compaction_2.CompactionPolicy; } });
Object.defineProperty(exports, "DynamicCompactionPolicy", { enumerable: true, get: function () { return compaction_2.DynamicCompactionPolicy; } });
Object.defineProperty(exports, "StaticCompactionPolicy", { enumerable: true, get: function () { return compaction_2.StaticCompactionPolicy; } });
Object.defineProperty(exports, "compaction", { enumerable: true, get: function () { return compaction_2.compaction; } });
var filesystem_2 = require("./filesystem.js");
Object.defineProperty(exports, "filesystem", { enumerable: true, get: function () { return filesystem_2.filesystem; } });
var memory_1 = require("./memory.js");
Object.defineProperty(exports, "memory", { enumerable: true, get: function () { return memory_1.memory; } });
var storage_1 = require("../memory/storage.js");
Object.defineProperty(exports, "InMemoryMemoryStore", { enumerable: true, get: function () { return storage_1.InMemoryMemoryStore; } });
var shell_2 = require("./shell.js");
Object.defineProperty(exports, "shell", { enumerable: true, get: function () { return shell_2.shell; } });
var skills_1 = require("./skills.js");
Object.defineProperty(exports, "skills", { enumerable: true, get: function () { return skills_1.skills; } });
function defaultCapabilities() {
    return [(0, filesystem_1.filesystem)(), (0, shell_1.shell)(), (0, compaction_1.compaction)()];
}
exports.Capabilities = {
    default: defaultCapabilities,
};
