// @openai/agents-core/dist/sandbox/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SANDBOX_ARCHIVE_LIMITS = void 0;
exports.normalizeSandboxClientCreateArgs = normalizeSandboxClientCreateArgs;
exports.resolveSandboxArchiveLimits = resolveSandboxArchiveLimits;
exports.validateSandboxArchiveLimits = validateSandboxArchiveLimits;
const manifest_1 = require("./manifest.js");
const typeGuards_1 = require("./shared/typeGuards.js");
const errors_1 = require("../errors.js");
exports.DEFAULT_SANDBOX_ARCHIVE_LIMITS = {
    maxInputBytes: 1024 * 1024 * 1024,
    maxExtractedBytes: 4 * 1024 * 1024 * 1024,
    maxMembers: 100_000,
};
function normalizeSandboxClientCreateArgs(args, manifestOptions) {
    if (args instanceof manifest_1.Manifest) {
        return {
            manifest: args,
            options: manifestOptions,
            snapshot: readSnapshotOption(manifestOptions),
            concurrencyLimits: readConcurrencyLimitsOption(manifestOptions),
            archiveLimits: readArchiveLimitsOption(manifestOptions),
        };
    }
    const manifest = args?.manifest;
    return {
        manifest: manifest
            ? manifest instanceof manifest_1.Manifest
                ? manifest
                : (0, manifest_1.cloneManifest)(manifest)
            : new manifest_1.Manifest(),
        options: args?.options,
        snapshot: args?.snapshot,
        concurrencyLimits: args?.concurrencyLimits,
        archiveLimits: args?.archiveLimits,
    };
}
function resolveSandboxArchiveLimits(limits) {
    if (limits == null) {
        return null;
    }
    validateSandboxArchiveLimits(limits);
    return {
        maxInputBytes: limits.maxInputBytes === undefined
            ? exports.DEFAULT_SANDBOX_ARCHIVE_LIMITS.maxInputBytes
            : limits.maxInputBytes,
        maxExtractedBytes: limits.maxExtractedBytes === undefined
            ? exports.DEFAULT_SANDBOX_ARCHIVE_LIMITS.maxExtractedBytes
            : limits.maxExtractedBytes,
        maxMembers: limits.maxMembers === undefined
            ? exports.DEFAULT_SANDBOX_ARCHIVE_LIMITS.maxMembers
            : limits.maxMembers,
    };
}
function validateSandboxArchiveLimits(limits) {
    if (limits == null) {
        return;
    }
    validatePositiveArchiveLimit('maxInputBytes', limits.maxInputBytes);
    validatePositiveArchiveLimit('maxExtractedBytes', limits.maxExtractedBytes);
    validatePositiveArchiveLimit('maxMembers', limits.maxMembers);
}
function validatePositiveArchiveLimit(name, value) {
    if (value == null) {
        return;
    }
    if (!Number.isSafeInteger(value) || value < 1) {
        throw new errors_1.UserError(`archiveLimits.${name} must be at least 1.`);
    }
}
function readSnapshotOption(options) {
    if (!(0, typeGuards_1.isRecord)(options)) {
        return undefined;
    }
    return options.snapshot;
}
function readConcurrencyLimitsOption(options) {
    if (!(0, typeGuards_1.isRecord)(options)) {
        return undefined;
    }
    return options.concurrencyLimits;
}
function readArchiveLimitsOption(options) {
    if (!(0, typeGuards_1.isRecord)(options)) {
        return undefined;
    }
    return options.archiveLimits;
}
