// @openai/agents-core/dist/sandbox/workspacePaths.mjs
import { UserError } from "../errors.mjs";
import { SandboxInvalidManifestPathError } from "./errors.mjs";
import { normalizePathGrant, } from "./pathGrants.mjs";
import { hasBackslashPathSeparator, hasEscapingParentPathSegment, hasParentPathSegment, isUnderPosixPath, normalizePosixPath, relativePosixPathWithinRoot, } from "./shared/posixPath.mjs";
export class WorkspacePathPolicy {
    root;
    extraPathGrants;
    constructor(options) {
        if (hasParentPathSegment(options.root)) {
            throw new UserError('Sandbox workspace root must not contain parent segments.');
        }
        if (hasBackslashPathSeparator(options.root)) {
            throw new UserError(`Sandbox path "${options.root}" must use "/" separators.`);
        }
        const root = normalizePosixPath(options.root);
        if (!root.startsWith('/')) {
            throw new UserError('Sandbox workspace root must be absolute.');
        }
        this.root = root;
        this.extraPathGrants = (options.extraPathGrants ?? [])
            .map((grant) => normalizePathGrant(grant))
            .sort((left, right) => right.path.length - left.path.length);
    }
    resolve(path, options = {}) {
        const originalPath = path ?? '';
        const trimmed = originalPath.trim();
        const normalizedPath = trimmed.length === 0
            ? this.root
            : trimmed.startsWith('/')
                ? normalizeSandboxPath(trimmed, originalPath)
                : joinPosixPath(this.root, normalizeSandboxPath(trimmed, originalPath));
        const workspaceRelativePath = relativePosixPathWithinRoot(this.root, normalizedPath);
        if (workspaceRelativePath !== null) {
            return {
                path: normalizedPath,
                workspaceRelativePath,
            };
        }
        const grant = this.matchingGrant(normalizedPath);
        if (grant) {
            if (options.forWrite && grant.readOnly) {
                throw new UserError(`Sandbox path "${originalPath}" uses read-only extra path grant "${grant.path}".`);
            }
            return {
                path: normalizedPath,
                grant,
            };
        }
        throw new SandboxInvalidManifestPathError(`Sandbox path "${originalPath}" escapes the workspace root.`);
    }
    matchingGrant(path) {
        return this.extraPathGrants.find((grant) => isUnderPosixPath(path, grant.path));
    }
}
function joinPosixPath(root, relativePath) {
    if (!relativePath) {
        return normalizePosixPath(root);
    }
    return normalizePosixPath(`${root.replace(/\/+$/u, '')}/${relativePath}`);
}
function normalizeSandboxPath(path, originalPath) {
    if (hasBackslashPathSeparator(path)) {
        throw new UserError(`Sandbox path "${originalPath}" must use "/" separators.`);
    }
    if (hasEscapingParentPathSegment(path)) {
        throw new SandboxInvalidManifestPathError(`Sandbox path "${originalPath}" must not escape root.`);
    }
    return normalizePosixPath(path);
}
