// @openai/agents-core/dist/sandbox/workspacePaths.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspacePathPolicy = void 0;
const errors_1 = require("../errors.js");
const errors_2 = require("./errors.js");
const pathGrants_1 = require("./pathGrants.js");
const posixPath_1 = require("./shared/posixPath.js");
class WorkspacePathPolicy {
    root;
    extraPathGrants;
    constructor(options) {
        if ((0, posixPath_1.hasParentPathSegment)(options.root)) {
            throw new errors_1.UserError('Sandbox workspace root must not contain parent segments.');
        }
        if ((0, posixPath_1.hasBackslashPathSeparator)(options.root)) {
            throw new errors_1.UserError(`Sandbox path "${options.root}" must use "/" separators.`);
        }
        const root = (0, posixPath_1.normalizePosixPath)(options.root);
        if (!root.startsWith('/')) {
            throw new errors_1.UserError('Sandbox workspace root must be absolute.');
        }
        this.root = root;
        this.extraPathGrants = (options.extraPathGrants ?? [])
            .map((grant) => (0, pathGrants_1.normalizePathGrant)(grant))
            .sort((left, right) => right.path.length - left.path.length);
    }
    resolve(path, options = {}) {
        const originalPath = path ?? '';
        const trimmed = originalPath.trim();
        const normalizedPath = trimmed.length === 0
            ? this.root
            : trimmed.startsWith('/')
                ? normalizeSandboxPath(trimmed, originalPath)
                : joinPosixPath(this.root, normalizeSandboxPath(trimmed, originalPath));
        const workspaceRelativePath = (0, posixPath_1.relativePosixPathWithinRoot)(this.root, normalizedPath);
        if (workspaceRelativePath !== null) {
            return {
                path: normalizedPath,
                workspaceRelativePath,
            };
        }
        const grant = this.matchingGrant(normalizedPath);
        if (grant) {
            if (options.forWrite && grant.readOnly) {
                throw new errors_1.UserError(`Sandbox path "${originalPath}" uses read-only extra path grant "${grant.path}".`);
            }
            return {
                path: normalizedPath,
                grant,
            };
        }
        throw new errors_2.SandboxInvalidManifestPathError(`Sandbox path "${originalPath}" escapes the workspace root.`);
    }
    matchingGrant(path) {
        return this.extraPathGrants.find((grant) => (0, posixPath_1.isUnderPosixPath)(path, grant.path));
    }
}
exports.WorkspacePathPolicy = WorkspacePathPolicy;
function joinPosixPath(root, relativePath) {
    if (!relativePath) {
        return (0, posixPath_1.normalizePosixPath)(root);
    }
    return (0, posixPath_1.normalizePosixPath)(`${root.replace(/\/+$/u, '')}/${relativePath}`);
}
function normalizeSandboxPath(path, originalPath) {
    if ((0, posixPath_1.hasBackslashPathSeparator)(path)) {
        throw new errors_1.UserError(`Sandbox path "${originalPath}" must use "/" separators.`);
    }
    if ((0, posixPath_1.hasEscapingParentPathSegment)(path)) {
        throw new errors_2.SandboxInvalidManifestPathError(`Sandbox path "${originalPath}" must not escape root.`);
    }
    return (0, posixPath_1.normalizePosixPath)(path);
}
