// @openai/agents-realtime/dist/bundle/agent-Do-OaBbR.mjs
var U = Object.defineProperty;
var j = (t, s, e) => s in t ? U(t, s, { enumerable: !0, configurable: !0, writable: !0, value: e }) : t[s] = e;
var o = (t, s, e) => j(t, typeof s != "symbol" ? s + "" : s, e);
import { Capability as _, requireBoundSession as N, UserError as b, tool as m, object as y, string as p, applyPatchTool as K, withSandboxSpan as w, isRecord as L, encodeUint8ArrayToBase64 as H, ApplyPatchOperation as z, number as d, boolean as P, prompt as T, Agent as X, SANDBOX_AGENT_BRAND as q, cloneManifest as G, Manifest as Y, normalizeUser as J } from "./index-D3rh1Df0.mjs";
function V(t) {
  return t ? !(typeof t == "object" && t && "constructor" in t && typeof t.constructor == "function" ? t.constructor.name : "").includes("ChatCompletions") : !0;
}
function Q(t) {
  return $(t);
}
function $(t) {
  return t ? !(typeof t == "object" && t && "constructor" in t && typeof t.constructor == "function" ? t.constructor.name : "").includes("ChatCompletions") : !1;
}
class Z {
}
class ee extends Z {
  constructor(e = 0.9, i = 24e4) {
    super();
    o(this, "thresholdRatio");
    o(this, "fallbackThreshold");
    this.thresholdRatio = e, this.fallbackThreshold = i;
  }
  compactThreshold(e) {
    if (!e)
      return this.fallbackThreshold;
    const i = u.forModel(e);
    return i ? Math.floor(i.contextWindow * this.thresholdRatio) : this.fallbackThreshold;
  }
}
class u {
  constructor(s) {
    o(this, "contextWindow");
    this.contextWindow = s;
  }
  static forModel(s) {
    for (const e of ie(s)) {
      if (ne.has(e))
        return new u(1047576);
      if (re.has(e))
        return new u(4e5);
      if (oe.has(e))
        return new u(2e5);
      if (ae.has(e))
        return new u(128e3);
    }
  }
}
class te extends _ {
  constructor(e = new ee()) {
    super();
    o(this, "type", "compaction");
    o(this, "policy");
    this.policy = e;
  }
  samplingParams(e) {
    if (e.modelInstance && !V(e.modelInstance))
      return {};
    const i = typeof e.model == "string" ? e.model : void 0;
    return {
      context_management: [
        {
          type: "compaction",
          compact_threshold: this.policy.compactThreshold(i)
        }
      ]
    };
  }
  processContext(e) {
    let i = -1;
    for (let n = 0; n < e.length; n += 1)
      e[n].type === "compaction" && (i = n);
    return i === -1 ? e : e.slice(i);
  }
}
function se(t = {}) {
  return new te(t.policy);
}
function ie(t) {
  const s = t.trim();
  if (!s)
    return [];
  const i = (s.includes("/") ? s.slice(s.lastIndexOf("/") + 1) : s).toLowerCase(), n = i.replace(/-\d{4}-\d{2}-\d{2}$/u, "");
  return [.../* @__PURE__ */ new Set([i, n])];
}
const ne = /* @__PURE__ */ new Set([
  "gpt-5.4",
  "gpt-5.4-2026-03-05",
  "gpt-5.4-pro",
  "gpt-5.4-pro-2026-03-05",
  "gpt-4.1",
  "gpt-4.1-2025-04-14",
  "gpt-4.1-mini",
  "gpt-4.1-mini-2025-04-14",
  "gpt-4.1-nano",
  "gpt-4.1-nano-2025-04-14"
]), re = /* @__PURE__ */ new Set([
  "gpt-5",
  "gpt-5-codex",
  "gpt-5-mini",
  "gpt-5-nano",
  "gpt-5-pro",
  "gpt-5.1",
  "gpt-5.1-codex",
  "gpt-5.1-codex-max",
  "gpt-5.1-codex-mini",
  "gpt-5.2",
  "gpt-5.2-codex",
  "gpt-5.2-codex-cyber",
  "gpt-5.2-pro",
  "gpt-5.3-codex",
  "gpt-5.3-codex-spark",
  "gpt-5.4-mini",
  "gpt-5.4-nano"
]), oe = /* @__PURE__ */ new Set([
  "codex-mini-latest",
  "o1",
  "o1-2024-12-17",
  "o1-pro",
  "o3",
  "o3-2025-04-16",
  "o3-mini",
  "o3-pro",
  "o3-deep-research",
  "o4-mini",
  "o4-mini-2025-04-16",
  "o4-mini-deep-research"
]), ae = /* @__PURE__ */ new Set([
  "gpt-4o",
  "gpt-4o-2024-08-06",
  "gpt-4o-mini",
  "gpt-4o-mini-2024-07-18",
  "gpt-5-chat-latest",
  "gpt-5.1-chat-latest",
  "gpt-5.2-chat-latest",
  "gpt-5.3-chat-latest"
]), le = {
  type: "object",
  properties: {
    patch: {
      type: "string",
      description: "A freeform apply_patch payload wrapped in *** Begin Patch and *** End Patch."
    },
    operation: {
      type: "object",
      description: "A single structured apply_patch operation."
    },
    operations: {
      type: "array",
      items: { type: "object" },
      description: "Structured apply_patch operations to apply in order."
    },
    command: {
      type: "array",
      description: 'Optional ["apply_patch", patch] command tuple form.'
    },
    type: {
      type: "string",
      enum: ["create_file", "update_file", "delete_file"],
      description: "The patch operation to apply."
    },
    path: {
      type: "string",
      description: "The sandbox file path to patch."
    },
    diff: {
      type: "string",
      description: "The V4A patch diff for create_file and update_file operations."
    },
    moveTo: {
      type: "string",
      description: "Optional destination path for update_file move operations."
    }
  },
  required: [],
  additionalProperties: !0
}, k = "*** Begin Patch", C = "*** End Patch", x = "*** Add File: ", I = "*** Delete File: ", v = "*** Update File: ", D = "*** Move to: ";
function A(t) {
  if (t instanceof Error)
    return t.message || t.toString();
  try {
    return JSON.stringify(t);
  } catch {
    return String(t);
  }
}
function ce(t) {
  if (typeof t == "string")
    return t;
  const s = t.image;
  if (typeof s == "string")
    return s;
  if (!L(s))
    return "No image data was returned by the sandbox session.";
  const e = s;
  return typeof e.url == "string" ? e.url : typeof e.fileId == "string" ? `OpenAI file reference: ${e.fileId}` : typeof e.data == "string" ? `data:${typeof e.mediaType == "string" ? e.mediaType : "application/octet-stream"};base64,${e.data}` : e.data instanceof Uint8Array ? `data:${typeof e.mediaType == "string" ? e.mediaType : "application/octet-stream"};base64,${H(e.data)}` : "No image data was returned by the sandbox session.";
}
function pe(t, s) {
  const e = A(s);
  if (/not found/iu.test(e))
    return `image path \`${t}\` was not found`;
  if (/not a file/iu.test(e))
    return `image path \`${t}\` is not a file`;
  if (/exceeds the 10 MB limit|exceeded the allowed size/iu.test(e))
    return `image path \`${t}\` exceeded the allowed size of 10MB; resize or compress the image and try again`;
  if (/unsupported image format|not a supported image/iu.test(e))
    return `image path \`${t}\` is not a supported image file`;
  const i = s instanceof Error ? s.name : typeof s;
  return `unable to read image at \`${t}\`: ${i}`;
}
function ue(t) {
  const s = t.trimStart();
  if (s.startsWith(k))
    return f(s);
  let e;
  try {
    e = JSON.parse(t);
  } catch (i) {
    return {
      ok: !1,
      error: `Invalid apply_patch JSON: ${A(i)}`
    };
  }
  return de(e);
}
function de(t) {
  if (typeof t == "string")
    return f(t);
  if (Array.isArray(t))
    return h(t);
  if (!L(t))
    return {
      ok: !1,
      error: "apply_patch input must be an object or array."
    };
  if (typeof t.patch == "string")
    return f(t.patch);
  if (Array.isArray(t.command)) {
    const [s, e] = t.command;
    if (s === "apply_patch" && typeof e == "string")
      return f(e);
  }
  return Array.isArray(t.operations) ? h(t.operations) : t.operation !== void 0 ? h([t.operation]) : h([t]);
}
function h(t) {
  const s = [];
  for (const e of t) {
    const i = z.safeParse(e);
    if (!i.success)
      return {
        ok: !1,
        error: `Invalid apply_patch operation: ${i.error.message}`
      };
    s.push(i.data);
  }
  return s.length === 0 ? { ok: !1, error: "apply_patch input must include an operation." } : { ok: !0, operations: s };
}
function f(t) {
  const s = t.split(/\r?\n/);
  if (s.at(-1) === "" && s.pop(), s[0] !== k)
    return {
      ok: !1,
      error: `apply_patch input must start with "${k}".`
    };
  if (s.length < 2 || s.at(-1) !== C)
    return {
      ok: !1,
      error: `apply_patch input must end with "${C}".`
    };
  const e = [];
  let i = 1;
  for (; i < s.length - 1; ) {
    const n = s[i];
    let r;
    if (n.startsWith(x))
      r = he(s, i);
    else if (n.startsWith(I))
      r = fe(s, i);
    else if (n.startsWith(v))
      r = me(s, i);
    else
      return {
        ok: !1,
        error: `Invalid apply_patch file operation header: ${n}`
      };
    if ("error" in r)
      return { ok: !1, error: r.error };
    e.push(r.operation), i = r.nextIndex;
  }
  return e.length === 0 ? {
    ok: !1,
    error: "apply_patch input must include at least one file operation."
  } : { ok: !0, operations: e };
}
function he(t, s) {
  const e = g(t[s], x);
  if (!e)
    return { error: `Missing path in apply_patch header: ${t[s]}` };
  s += 1;
  const i = [];
  for (; s < t.length - 1 && !S(t[s]); ) {
    const n = t[s];
    if (!n.startsWith("+"))
      return { error: `Invalid Add File line: ${n}` };
    i.push(n), s += 1;
  }
  return i.length === 0 ? {
    error: `Add File patch for ${e} must include at least one + line.`
  } : {
    operation: {
      type: "create_file",
      path: e,
      diff: F(i)
    },
    nextIndex: s
  };
}
function fe(t, s) {
  const e = g(t[s], I);
  return e ? (s += 1, s < t.length - 1 && !S(t[s]) ? { error: `Delete File patch for ${e} must not include a diff.` } : {
    operation: {
      type: "delete_file",
      path: e
    },
    nextIndex: s
  }) : { error: `Missing path in apply_patch header: ${t[s]}` };
}
function me(t, s) {
  const e = g(t[s], v);
  if (!e)
    return { error: `Missing path in apply_patch header: ${t[s]}` };
  s += 1;
  let i;
  if (s < t.length - 1 && t[s].startsWith(D)) {
    const r = g(t[s], D);
    if (!r)
      return { error: `Missing path in apply_patch header: ${t[s]}` };
    i = r, s += 1;
  }
  const n = [];
  for (; s < t.length - 1 && !S(t[s]); )
    n.push(t[s]), s += 1;
  return n.length === 0 && !i ? { error: `Update File patch for ${e} must include a hunk.` } : {
    operation: {
      type: "update_file",
      path: e,
      diff: n.length > 0 ? F(n) : "",
      ...i ? { moveTo: i } : {}
    },
    nextIndex: s
  };
}
function g(t, s) {
  const e = t.slice(s.length).trim();
  return e || null;
}
function S(t) {
  return t.startsWith(x) || t.startsWith(I) || t.startsWith(v);
}
function F(t) {
  return `${t.join(`
`)}
`;
}
async function ye(t, s, e) {
  let i;
  try {
    switch (s.type) {
      case "create_file":
        i = await t.createFile(s, { runContext: e });
        break;
      case "update_file":
        i = await t.updateFile(s, { runContext: e });
        break;
      case "delete_file":
        i = await t.deleteFile(s, { runContext: e });
        break;
      default:
        return "Unsupported apply_patch operation.";
    }
  } catch (n) {
    return `Failed to apply patch: ${A(n)}`;
  }
  return (i == null ? void 0 : i.status) === "failed" ? i.output ? `Patch failed: ${i.output}` : "Patch failed." : (i == null ? void 0 : i.output) ?? "Patch applied.";
}
async function ge(t, s, e) {
  const i = [];
  for (const n of s) {
    const r = await ye(t, n, e);
    r && i.push(r);
  }
  return i.join(`
`) || "Patch applied.";
}
function be(t) {
  return {
    type: "function",
    name: "apply_patch",
    description: "Applies a create, update, move, or delete file patch in the sandbox workspace. Accepts a freeform patch string or structured apply_patch operations.",
    parameters: le,
    strict: !1,
    deferLoading: !1,
    needsApproval: async () => !1,
    isEnabled: async () => !0,
    invoke: async (e, i) => {
      const n = ue(i);
      return n.ok ? await ge(t, n.operations, e) : n.error;
    }
  };
}
class we extends _ {
  constructor(e = {}) {
    super();
    o(this, "type", "filesystem");
    o(this, "configureTools");
    this.configureTools = e.configureTools;
  }
  tools() {
    var l;
    const e = N(this.type, this._session), i = (l = e.createEditor) == null ? void 0 : l.call(e, this._runAs);
    if (!i)
      throw new b("Filesystem sandbox sessions must provide createEditor().");
    const n = [], r = async (a) => {
      if (!e.viewImage)
        throw new b("Filesystem sandbox sessions must provide viewImage().");
      try {
        return await w("sandbox.view_image", {
          path: a,
          run_as: this._runAs
        }, async () => await e.viewImage({
          path: a,
          runAs: this._runAs
        }));
      } catch (c) {
        return pe(a, c);
      }
    };
    return $(this._modelInstance) ? n.push(m({
      name: "view_image",
      description: "Returns an image output from a local path in the sandbox workspace.",
      parameters: y({
        path: p().describe("Local filesystem path to an image file")
      }),
      execute: async ({ path: a }) => await r(a)
    })) : n.push(m({
      name: "view_image",
      description: "Returns an image from a local path in the sandbox workspace as a data URL or reference string.",
      parameters: y({
        path: p().describe("Local filesystem path to an image file")
      }),
      execute: async ({ path: a }) => ce(await r(a))
    })), Q(this._modelInstance) ? n.push(K({
      name: "apply_patch",
      editor: i
    })) : n.push(be(i)), this.configureTools ? this.configureTools([...n]) : n;
  }
}
function ke(t = {}) {
  return new we(t);
}
const _e = 1e4, Te = 250, xe = [
  "When using the shell:",
  "- Use `exec_command` for shell execution.",
  "- If available, use `write_stdin` to interact with or poll running sessions.",
  "- To interrupt a long-running process via `write_stdin`, start it with `tty=true` and send Ctrl-C (`\\u0003`).",
  "- Prefer `rg` and `rg --files` for text/file discovery when available.",
  "- Avoid using Python scripts just to print large file chunks."
].join(`
`);
class Ie extends _ {
  constructor(e = {}) {
    super();
    o(this, "type", "shell");
    o(this, "configureTools");
    this.configureTools = e.configureTools;
  }
  tools() {
    var n;
    const e = N(this.type, this._session);
    if (!e.execCommand)
      throw new b("Shell sandbox sessions must provide execCommand().");
    const i = [
      m({
        name: "exec_command",
        description: "Runs a command in a PTY, returning output or a session ID for ongoing interaction.",
        parameters: y({
          cmd: p().min(1).describe("Shell command to execute."),
          workdir: p().optional().describe("Optional working directory to run the command in; defaults to the turn cwd."),
          shell: p().optional().describe("Shell binary to launch. Defaults to the user's default shell."),
          login: P().default(!0).describe("Whether to run the shell with -l/-i semantics. Defaults to true."),
          tty: P().default(!1).describe("Whether to allocate a TTY for the command. Defaults to false (plain pipes); set to true to open a PTY and access TTY process."),
          yield_time_ms: d().int().min(0).default(_e).describe("How long to wait (in milliseconds) for output before yielding."),
          max_output_tokens: d().int().min(1).optional().describe("Maximum number of tokens to return. Excess output will be truncated.")
        }),
        execute: async ({ cmd: r, workdir: l, shell: a, login: c, tty: E, yield_time_ms: B, max_output_tokens: R }) => await w("sandbox.exec", {
          cmd: r,
          workdir: l,
          shell: a,
          login: c,
          tty: E,
          run_as: this._runAs
        }, async () => await e.execCommand({
          cmd: r,
          workdir: l,
          shell: a,
          login: c,
          tty: E,
          yieldTimeMs: B,
          maxOutputTokens: R,
          runAs: this._runAs
        }))
      })
    ];
    return (n = e.supportsPty) != null && n.call(e) && e.writeStdin && i.push(m({
      name: "write_stdin",
      description: "Writes characters to an existing unified exec session and returns recent output.",
      parameters: y({
        session_id: d().int().describe("Identifier of the running unified exec session."),
        chars: p().default("").describe("Bytes to write to stdin (may be empty to poll)."),
        yield_time_ms: d().int().min(0).default(Te).describe("How long to wait (in milliseconds) for output before yielding."),
        max_output_tokens: d().int().min(1).optional().describe("Maximum number of tokens to return. Excess output will be truncated.")
      }),
      execute: async ({ session_id: r, chars: l, yield_time_ms: a, max_output_tokens: c }) => await w("sandbox.write_stdin", {
        session_id: r
      }, async () => await e.writeStdin({
        sessionId: r,
        chars: l,
        yieldTimeMs: a,
        maxOutputTokens: c
      }))
    })), this.configureTools ? this.configureTools([...i]) : i;
  }
  instructions() {
    return xe;
  }
}
function ve(t = {}) {
  return new Ie(t);
}
T`
A skill is a set of local instructions to follow that is stored in a \`SKILL.md\` file. Below is the list of skills that can be used. Each entry includes a name, description, and file path so you can open the source for full instructions when using a specific skill.
`;
T`
### How to use skills
- Discovery: The list above is the skills available in this session (name + description + file path). Skill bodies live on disk at the listed paths.
- Trigger rules: If the user names a skill (with \`$SkillName\` or plain text) OR the task clearly matches a skill's description shown above, you must use that skill for that turn. Multiple mentions mean use them all. Do not carry skills across turns unless re-mentioned.
- Missing/blocked: If a named skill isn't in the list or the path can't be read, say so briefly and continue with the best fallback.
- How to use a skill (progressive disclosure):
  1) After deciding to use a skill, open its \`SKILL.md\`. Read only enough to follow the workflow.
  2) If \`SKILL.md\` points to extra folders such as \`references/\`, load only the specific files needed for the request; don't bulk-load everything.
  3) If \`scripts/\` exist, prefer running or patching them instead of retyping large code blocks.
  4) If \`assets/\` or templates exist, reuse them instead of recreating from scratch.
- Coordination and sequencing:
  - If multiple skills apply, choose the minimal set that covers the request and state the order you'll use them.
  - Announce which skill(s) you're using and why (one short line). If you skip an obvious skill, say why.
- Context hygiene:
  - Keep context small: summarize long sections instead of pasting them; only load extra files when needed.
  - Avoid deep reference-chasing: prefer opening only files directly linked from \`SKILL.md\` unless you're blocked.
  - When variants exist (frameworks, providers, domains), pick only the relevant reference file(s) and note that choice.
- Safety and fallback: If a skill can't be applied cleanly (missing files, unclear instructions), state the issue, pick the next-best approach, and continue.
`;
T`
### How to use skills
- Discovery: The list above is the skill index available in this session (name + description + workspace path). In lazy mode, those paths are loaded on demand instead of being present up front.
- Trigger rules: If the user names a skill (with \`$SkillName\` or plain text) OR the task clearly matches a skill's description shown above, you must use that skill for that turn. Multiple mentions mean use them all. Do not carry skills across turns unless re-mentioned.
- Missing/blocked: If a named skill isn't in the list or the path can't be read, say so briefly and continue with the best fallback.
- How to use a skill (progressive disclosure):
  1) After deciding to use a lazy skill, call \`load_skill\` for that skill first, then open its \`SKILL.md\`.
  2) If \`SKILL.md\` points to extra folders such as \`references/\`, load only the specific files needed for the request; don't bulk-load everything.
  3) If \`scripts/\` exist, prefer running or patching them instead of retyping large code blocks.
  4) If \`assets/\` or templates exist, reuse them instead of recreating from scratch.
- Coordination and sequencing:
  - If multiple skills apply, choose the minimal set that covers the request and state the order you'll use them.
  - Announce which skill(s) you're using and why (one short line). If you skip an obvious skill, say why.
- Context hygiene:
  - Keep context small: summarize long sections instead of pasting them; only load extra files when needed.
  - Avoid deep reference-chasing: prefer opening only files directly linked from \`SKILL.md\` unless you're blocked.
  - When variants exist (frameworks, providers, domains), pick only the relevant reference file(s) and note that choice.
- Safety and fallback: If a skill can't be applied cleanly (missing files, unclear instructions), state the issue, pick the next-best approach, and continue.
`;
function Ae() {
  return [ke(), ve(), se()];
}
const Se = {
  default: Ae
};
var M, O;
class W extends (O = X, M = q, O) {
  constructor(e) {
    super(e);
    o(this, M, !0);
    o(this, "defaultManifest");
    o(this, "baseInstructions");
    o(this, "capabilities");
    o(this, "runAs");
    o(this, "runtimeManifest");
    if (e.baseInstructions !== void 0 && typeof e.baseInstructions != "string" && typeof e.baseInstructions != "function")
      throw new TypeError("SandboxAgent baseInstructions must be a string or function.");
    this.defaultManifest = e.defaultManifest ? G(e.defaultManifest) : void 0, this.baseInstructions = e.baseInstructions, this.capabilities = e.capabilities ?? Se.default(), this.runAs = Ee(e.runAs), this.runtimeManifest = this.defaultManifest ?? new Y();
  }
  clone(e) {
    return new W({
      name: e.name ?? this.name,
      instructions: e.instructions ?? this.instructions,
      prompt: e.prompt ?? this.prompt,
      handoffDescription: e.handoffDescription ?? this.handoffDescription,
      handoffs: e.handoffs ?? this.handoffs,
      model: e.model ?? this.model,
      modelSettings: e.modelSettings ?? this.modelSettings,
      tools: e.tools ?? this.tools,
      mcpServers: e.mcpServers ?? this.mcpServers,
      mcpConfig: e.mcpConfig ?? this.mcpConfig,
      inputGuardrails: e.inputGuardrails ?? this.inputGuardrails,
      outputGuardrails: e.outputGuardrails ?? this.outputGuardrails,
      outputType: e.outputType ?? this.outputType,
      toolUseBehavior: e.toolUseBehavior ?? this.toolUseBehavior,
      resetToolChoice: e.resetToolChoice ?? this.resetToolChoice,
      defaultManifest: e.defaultManifest ?? this.defaultManifest,
      baseInstructions: e.baseInstructions ?? this.baseInstructions,
      capabilities: e.capabilities ?? this.capabilities,
      runAs: e.runAs ?? this.runAs
    });
  }
}
function Ee(t) {
  if (t !== void 0) {
    if (typeof t == "string") {
      const s = t.trim();
      if (!s)
        throw new TypeError("SandboxAgent runAs must be non-empty.");
      return s;
    }
    return J(t);
  }
}
export {
  W as SandboxAgent
};
