// @openai/agents-core/dist/sandbox/shared/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.materializeEnvironment = materializeEnvironment;
exports.materializeStaticEnvironment = materializeStaticEnvironment;
exports.mergeMaterializedEnvironment = mergeMaterializedEnvironment;
exports.mergeStaticMaterializedEnvironment = mergeStaticMaterializedEnvironment;
exports.serializeManifestEnvironment = serializeManifestEnvironment;
exports.serializeRuntimeEnvironmentForPersistence = serializeRuntimeEnvironmentForPersistence;
exports.deserializePersistedEnvironmentForRuntime = deserializePersistedEnvironmentForRuntime;
exports.assertShellEnvironmentName = assertShellEnvironmentName;
const errors_1 = require("../../errors.js");
const SHELL_ENV_NAME_PATTERN = /^[A-Za-z_][A-Za-z0-9_]*$/;
async function materializeEnvironment(manifest, baseEnvironment = {}) {
    return {
        ...baseEnvironment,
        ...(await manifest.resolveEnvironment()),
    };
}
function materializeStaticEnvironment(manifest) {
    return Object.fromEntries(Object.entries(manifest.environment).map(([key, value]) => [
        key,
        value.value,
    ]));
}
async function mergeMaterializedEnvironment(previousManifest, nextManifest, currentEnvironment) {
    const previousEnvironment = await materializeEnvironment(previousManifest);
    const runtimeEnvironmentDelta = Object.fromEntries(Object.entries(currentEnvironment).filter(([key, value]) => previousEnvironment[key] !== value));
    return await materializeEnvironment(nextManifest, runtimeEnvironmentDelta);
}
function mergeStaticMaterializedEnvironment(previousManifest, nextManifest, currentEnvironment) {
    const previousEnvironment = materializeStaticEnvironment(previousManifest);
    const runtimeEnvironmentDelta = Object.fromEntries(Object.entries(currentEnvironment).filter(([key, value]) => previousEnvironment[key] !== value));
    return {
        ...runtimeEnvironmentDelta,
        ...materializeStaticEnvironment(nextManifest),
    };
}
function serializeManifestEnvironment(manifest) {
    return Object.fromEntries(Object.entries(manifest.environment).map(([key, value]) => [
        key,
        value.normalized(),
    ]));
}
function serializeRuntimeEnvironmentForPersistence(manifest, environment) {
    return Object.fromEntries(Object.entries(manifest.environment)
        .filter(([key, value]) => !value.ephemeral &&
        key in environment &&
        typeof environment[key] === 'string')
        .map(([key, value]) => [
        key,
        value.resolver ? environment[key] : value.value,
    ]));
}
function deserializePersistedEnvironmentForRuntime(manifest, environment, baseEnvironment = {}) {
    const manifestEnvironment = materializeStaticEnvironment(manifest);
    const persistentManifestEnvironment = Object.fromEntries(Object.entries(manifestEnvironment).filter(([key]) => !manifest.environment[key]?.ephemeral));
    const persistedEnvironment = Object.fromEntries(Object.entries(environment ?? {}).filter(([key]) => key in manifestEnvironment && !manifest.environment[key]?.ephemeral));
    return {
        ...baseEnvironment,
        ...persistentManifestEnvironment,
        ...persistedEnvironment,
    };
}
function assertShellEnvironmentName(name) {
    if (!SHELL_ENV_NAME_PATTERN.test(name)) {
        throw new errors_1.UserError(`Invalid environment variable name "${name}". Environment names used in shell commands must match ${SHELL_ENV_NAME_PATTERN.source}.`);
    }
}
