// @openai/agents-core/dist/sandbox/shared/environment.mjs
import { UserError } from "../../errors.mjs";
const SHELL_ENV_NAME_PATTERN = /^[A-Za-z_][A-Za-z0-9_]*$/;
export async function materializeEnvironment(manifest, baseEnvironment = {}) {
    return {
        ...baseEnvironment,
        ...(await manifest.resolveEnvironment()),
    };
}
export function materializeStaticEnvironment(manifest) {
    return Object.fromEntries(Object.entries(manifest.environment).map(([key, value]) => [
        key,
        value.value,
    ]));
}
export async function mergeMaterializedEnvironment(previousManifest, nextManifest, currentEnvironment) {
    const previousEnvironment = await materializeEnvironment(previousManifest);
    const runtimeEnvironmentDelta = Object.fromEntries(Object.entries(currentEnvironment).filter(([key, value]) => previousEnvironment[key] !== value));
    return await materializeEnvironment(nextManifest, runtimeEnvironmentDelta);
}
export function mergeStaticMaterializedEnvironment(previousManifest, nextManifest, currentEnvironment) {
    const previousEnvironment = materializeStaticEnvironment(previousManifest);
    const runtimeEnvironmentDelta = Object.fromEntries(Object.entries(currentEnvironment).filter(([key, value]) => previousEnvironment[key] !== value));
    return {
        ...runtimeEnvironmentDelta,
        ...materializeStaticEnvironment(nextManifest),
    };
}
export function serializeManifestEnvironment(manifest) {
    return Object.fromEntries(Object.entries(manifest.environment).map(([key, value]) => [
        key,
        value.normalized(),
    ]));
}
export function serializeRuntimeEnvironmentForPersistence(manifest, environment) {
    return Object.fromEntries(Object.entries(manifest.environment)
        .filter(([key, value]) => !value.ephemeral &&
        key in environment &&
        typeof environment[key] === 'string')
        .map(([key, value]) => [
        key,
        value.resolver ? environment[key] : value.value,
    ]));
}
export function deserializePersistedEnvironmentForRuntime(manifest, environment, baseEnvironment = {}) {
    const manifestEnvironment = materializeStaticEnvironment(manifest);
    const persistentManifestEnvironment = Object.fromEntries(Object.entries(manifestEnvironment).filter(([key]) => !manifest.environment[key]?.ephemeral));
    const persistedEnvironment = Object.fromEntries(Object.entries(environment ?? {}).filter(([key]) => key in manifestEnvironment && !manifest.environment[key]?.ephemeral));
    return {
        ...baseEnvironment,
        ...persistentManifestEnvironment,
        ...persistedEnvironment,
    };
}
export function assertShellEnvironmentName(name) {
    if (!SHELL_ENV_NAME_PATTERN.test(name)) {
        throw new UserError(`Invalid environment variable name "${name}". Environment names used in shell commands must match ${SHELL_ENV_NAME_PATTERN.source}.`);
    }
}
