// @openai/agents-core/dist/sandbox/shared/media.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_VIEW_IMAGE_BYTES = void 0;
exports.assertViewImageByteLength = assertViewImageByteLength;
exports.imageOutputFromBytes = imageOutputFromBytes;
exports.sniffImageMediaType = sniffImageMediaType;
exports.toUint8Array = toUint8Array;
const errors_1 = require("../../errors.js");
exports.MAX_VIEW_IMAGE_BYTES = 10 * 1024 * 1024;
function assertViewImageByteLength(path, byteLength) {
    if (byteLength > exports.MAX_VIEW_IMAGE_BYTES) {
        throw new errors_1.UserError(`Image file exceeds the 10 MB limit: ${path}`);
    }
}
function imageOutputFromBytes(path, bytes) {
    assertViewImageByteLength(path, bytes.byteLength);
    const mediaType = sniffImageMediaType(bytes, path);
    if (!mediaType) {
        throw new errors_1.UserError(`Unsupported image format for view_image: ${path}`);
    }
    return {
        type: 'image',
        image: {
            data: Uint8Array.from(bytes),
            mediaType,
        },
    };
}
function sniffImageMediaType(bytes, path) {
    if (bytes[0] === 0x89 &&
        bytes[1] === 0x50 &&
        bytes[2] === 0x4e &&
        bytes[3] === 0x47) {
        return 'image/png';
    }
    if (bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
        return 'image/jpeg';
    }
    if (bytes[0] === 0x47 &&
        bytes[1] === 0x49 &&
        bytes[2] === 0x46 &&
        bytes[3] === 0x38) {
        return 'image/gif';
    }
    if (bytes[0] === 0x52 &&
        bytes[1] === 0x49 &&
        bytes[2] === 0x46 &&
        bytes[3] === 0x46 &&
        bytes[8] === 0x57 &&
        bytes[9] === 0x45 &&
        bytes[10] === 0x42 &&
        bytes[11] === 0x50) {
        return 'image/webp';
    }
    if (bytes[0] === 0x42 && bytes[1] === 0x4d) {
        return 'image/bmp';
    }
    if ((bytes[0] === 0x49 &&
        bytes[1] === 0x49 &&
        bytes[2] === 0x2a &&
        bytes[3] === 0x00) ||
        (bytes[0] === 0x4d &&
            bytes[1] === 0x4d &&
            bytes[2] === 0x00 &&
            bytes[3] === 0x2a)) {
        return 'image/tiff';
    }
    if (looksLikeSvg(bytes)) {
        return 'image/svg+xml';
    }
    return mediaTypeFromPath(path);
}
async function toUint8Array(value) {
    if (!value) {
        return new Uint8Array();
    }
    if (value instanceof Uint8Array) {
        return Uint8Array.from(value);
    }
    if (value instanceof ArrayBuffer) {
        return new Uint8Array(value);
    }
    if (ArrayBuffer.isView(value)) {
        return new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
    }
    if (typeof value === 'string') {
        return new TextEncoder().encode(value);
    }
    if (typeof Blob !== 'undefined' && value instanceof Blob) {
        return new Uint8Array(await value.arrayBuffer());
    }
    return Uint8Array.from(value);
}
function looksLikeSvg(bytes) {
    const prefix = new TextDecoder()
        .decode(bytes.subarray(0, Math.min(bytes.byteLength, 512)))
        .trimStart()
        .toLowerCase();
    return prefix.startsWith('<svg') || /^<\?xml[\s\S]*<svg/u.test(prefix);
}
function mediaTypeFromPath(path) {
    const lowerPath = path?.trim().toLowerCase() ?? '';
    if (lowerPath.endsWith('.png')) {
        return 'image/png';
    }
    if (lowerPath.endsWith('.jpg') || lowerPath.endsWith('.jpeg')) {
        return 'image/jpeg';
    }
    if (lowerPath.endsWith('.gif')) {
        return 'image/gif';
    }
    if (lowerPath.endsWith('.webp')) {
        return 'image/webp';
    }
    if (lowerPath.endsWith('.bmp')) {
        return 'image/bmp';
    }
    if (lowerPath.endsWith('.tif') || lowerPath.endsWith('.tiff')) {
        return 'image/tiff';
    }
    if (lowerPath.endsWith('.svg') || lowerPath.endsWith('.svgz')) {
        return 'image/svg+xml';
    }
    return null;
}
