// @openai/agents-core/dist/sandbox/shared/typeGuards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRecord = isRecord;
exports.isStringRecord = isStringRecord;
exports.readOptionalString = readOptionalString;
exports.readString = readString;
exports.readOptionalNumberArray = readOptionalNumberArray;
exports.readStringArray = readStringArray;
function isRecord(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
function isStringRecord(value) {
    return (isRecord(value) &&
        Object.values(value).every((entry) => typeof entry === 'string'));
}
function readOptionalString(value, key) {
    const result = value[key];
    return typeof result === 'string' ? result : undefined;
}
function readString(value, key, fallback = '') {
    const result = value[key];
    return result === undefined || result === null ? fallback : String(result);
}
function readOptionalNumberArray(value) {
    return Array.isArray(value)
        ? value.filter((entry) => typeof entry === 'number')
        : undefined;
}
function readStringArray(value) {
    return Array.isArray(value)
        ? value.filter((entry) => typeof entry === 'string')
        : [];
}
